---
title: "msigdb_collections"
author: "yincy"
date: "8/22/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# H collection: hallmark gene sets
We envision this collection as the starting point for your exploration of the MSigDB resource and GSEA. Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression. These gene sets were generated by a computational methodology based on identifying gene set overlaps and retaining genes that display coordinate expression. The hallmarks reduce noise and redundancy and provide a better delineated biological space for GSEA. We refer to the original overlapping gene sets, from which a hallmark is derived, as its 'founder' sets. Hallmark gene set pages provide links to the corresponding founder sets for deeper follow up.  

# C1 collection: position gene sets
Gene annotations for this collection are derived from the Chromosome and Karyotype band from Ensembl Biomart and reflect the gene architecture as represented on the primary assembly. Decimals in cytogenetic bands were ignored. For example, 5q31.1 was considered 5q31. Therefore, genes annotated as 5q31.2 and those annotated as 5q31.3 were both placed in the same set, 5q31. These gene sets can b helpful in identifying effects releated to chromosomal deletions or amplifications, dosage compensation, epigenetic silencing, and other regional effects.

# C2 collection: curated gene sets
Gene sets in this collection are curated from various sources, including online pathway databases and the biomedical literature. Many sets are also contributed by individual domain experts. The gene set page for each gene set lists its source. The C2 collection is divided into the following two sub-collections: Chemical an genetic perturbations (CGP) and Canonical pathways (CP).

## C2 sub-collection CGP: Chemical and genetic perturbations
Gene sets that represent expression signatures of genetic and chemical perturbations.

Most of the CGP sets came from the biomedical literature. Over the past several years, microarray studies have identified signatures of several important biological and clinical states (e.g. cancer metastasis, stem cell characteristics, drug resistance). The C2 collection makes many of these signatures, originally published as tables in a paper, available as gene sets. To do this, we compiled a list of microarray articles with published gene expression signatures and, from each article, extracted one or more gene sets from tables in the main text or supplementary information. A number of these gene sets come in pairs: xxx_UP (and xxx_DN) gene sets representing genes induced (and repressed) by the perturbation. The majority of CGP sets were curated from publications. They include links to the PubMed citation, the exact source of the set (e.g., Table 1), and links to any corresponding raw data in GEO or ArrayExpress repositories. When the gene set involves a genetic perturbation, the set's brief description includes a link to the gene's entry in the NCBI (Entrez) Gene database. When the gene set involves a chemical perturbation, the set's brief description includes a link to the chemical's entry in the NCBI PubChem Compound database.

Other CGP gene sets include:
- Gene sets contributed by the L2L database of published microarray gene expression data at University of Washington. See Newman JC, Weiner AM. L2L: a simple tool for discovering the hidden significance in microarray expression data. Genome Biol. 2005;6(9):R81. See also http://depts.washington.edu/l2l.

- Gene sets curated by Dr. Chi Dang from the MYC Target Gene Database at Johns Hopkins University School of Medicine. See Zeller KI, Jegga AG, Aronow BJ, O'Donnell KA, Dang CV. An integrated database of genes responsive to the Myc oncogenic transcription factor: identification of direct genomic targets. Genome Biol. 2003;4(10):R69.

- A number of individuals have contributed gene sets to this collection. The gene set annotation includes a "contributor" field that acknowledges the contributor by name/affiliation.


## C2 sub-collection CP: Canonical pathways
The pathway gene sets are curated from the following online databases:

- BioCarta： http://cgap.nci.nih.gov/Pathways/BioCarta_Pathways
- KEGG：http://www.pathway.jp
- Matrisome Project：http://matrisomeproject.mit.edu
- Pathway Interaction database：http://www.ndexbio.org)
- Reactome：http://www.reactome.org
- SigmaAldrich：http://www.sigmaaldrich.com/life-science.html
- Signaling Gateway：http://www.signaling-gateway.org
- SuperArray SABioscience：http://www.sabiosciences.com/ArrayList.php
- WikiPathways：https://www.wikipathways.org/


# C3 collection:regulatory target gene sets
Gene sets representing potential targets of regulation by transcription factors or microRNAs. The sets consist of genes grouped by their shared regulatory element. The motifs represent known or likely cis-regulatory elements in promoters and 3'-UTRs. These gene sets make it possible to link changes in an expression profiling experiment to a putative cis-regulatory element. The C3 collection is divided into two sub-collections: microRNA targets (MIR) and transcription factor targets (TFT).

## C3 sub-collection MIR: microRNA targets
- miRDB: These sets consist of computationally predicted human gene targets of miRNAs using the MirTarget algorith(Liu and Wang,2019). Data was curated from miRDB v6.0(mirdb.org, Chen and Wang, 2020) target predictions with MirTarget scores > 80(high confidence predictions). miRNAs catalogued in miRDB v6.0 are derived from miRBase v22(March 2018).

- MIR_legacy: (These are older gene sets that formerly represented the C3:MIR sub-collection prior to MSigDB v7.1).These sets consist of genes sharing 7-nucleotide motifs in their 3' untranslated regions. Each 7-mer motif matches (is complementary to) the seed (bases 2 through 8) of the mature human microRNA (miRNAs) catalogued in v7.1 of miRBase (October 2005).

## C3 sub-collection TFT:Transcription factor targets
- Sets of human genes predicted to contain transcription factor binding sites in their promoter regions (-1000,+100 bp around the transcription start site) for the indicated transcription factor. Gene sets are derived from the Gene Transcription Regulation Database (GTRD, gtrd.biouml.org) uniform processing pipeline and represent a candidate list of potential regulatory targets for each transcription factor (see MSigDB release notes for the current included GTRD version).

- TFT_Legacy:(These are older gene sets that formerly represented the C3:TFT sub-collection prior to MSigDB v7.1). Gene sets that share upstream cis-regulatory motifs which can function as potential transcription factor binding sites. We used two approaches to generate these motif gene sets.
    + Gene sets of "conserved instances" consist of the inferred target genes for each motif m of 174 motifs highly conserved in promoters of four mammalian species (human, mouse, rat and dog). The motifs represent potential transcription factor binding sites and are catalogued in Xie X, Lu J, Kulbokas EJ, Golub TR, Mootha V, Lindblad-Toh K, Lander ES, Kellis M. Systematic discovery of regulatory motifs in human promoters and 3' UTRs by comparison of several mammals. Nature. 2005 Mar 17;434(7031):338-45.. Each gene set consists of all human genes whose promoters contained at least one conserved instance of motif m , where a promoter is defined as the non-coding sequence contained within a 4-kilobase window centered at the transcription start site (TSS).
    + Mammalian transcriptional regulatory motifs were extracted from v7.4 TRANSFAC database (see supplementary data of Xie et al). Each gene set consists of all human genes whose promoters contains at least one conserved instance of the TRANSFAC motif, where a promoter is defined as the non-coding sequence contained within a 4-kilobase window centered at the transcription start site (TSS).
























