---
title: "Quick_start_to_harmony"
author: "YinCY"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
Harmony is an algorithm for performing integration of single cell genomics datasets.  

```{r}
knitr::include_graphics("figures/harmony.jpg")
```


# Integrating cell line datasets from 10X
We library normalized the cells, log transformed the counts, and scaled the genes. Then we performed PCA and kept the top 20 PCs. The PCA embeddings and meta data are available as part of this package.  

```{r}
library(harmony)
data('cell_lines')

V <- cell_lines$scaled_pcs
meta_data <- cell_lines$meta_data
```

Initially, the cells cluster by both dataset (left) and cell type (right).  

```{r}
library(ggplot2)
source("R/do_scater.R")
p1 <- do_scatter(V, meta_data = meta_data, label_name = "dataset") +
    labs(title = "Colored by dataset")
p2 <- do_scatter(V, meta_data, "cell_type") +
    labs(title = "Colored by cell type")

library(patchwork)
p1 + p2
```

Letâ€™s run `Harmony` to remove the influence of dataset-of-origin from the embedding. By default, `Harmony` accepts a normalized gene expression matrix and performs PCA. Since here we already have the PCs, we specify `do_pca=FALSE`. The matrix harmony_embeddings is the matrix of `Harmony` corrected PCA embeddings.

```{r}
harmony_embeddings <- HarmonyMatrix(data_mat = V, 
                                    meta_data = meta_data, 
                                    vars_use = "dataset", 
                                    do_pca = FALSE, 
                                    verbose = FALSE)
```

After Harmony, the datasets are now mixed (left) and the cell types are still separate (right).  

```{r}
p1 <- do_scatter(harmony_embeddings, 
                 meta_data = meta_data, 
                 label_name = "dataset", 
                 point_size = 1) +
    labs(title = "Colored by dataset")
p1
```


# Next Steps
## Interfacing to software packages
You can also run Harmony as part of an established pipeline in several packages, such as Seurat, MUDAN, and scran.

The most common way to run Harmony is on reduced dimensions such as PC embeddings from pricipal component analysis (PCA). If you use low dimensiional embeddings, set `do_pca = FALSE`. 

## Normalized gene matrix
You can also run Harmony on a sparse matrix of library size normalized expression counts. The `HarmonyMatrix()` function will scale expression data, run PCA, and run the Harmony integration algorithm.













