---
title: "02_single_cell"
author: "YinCY"
date: "2024/5/16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(magrittr)
library(GEOquery)
library(TCGAbiolinks)
library(SummarizedExperiment)

# getGEOSuppFiles(GEO = "GSE15459", baseDir = "data/")
```

```{r}
getGDCprojects()
getProjectSummary(project = "TCGA-STAD")

query_tcga_gc <- GDCquery(project = "TCGA-STAD", 
                          data.category = "Transcriptome Profiling", 
                          data.type = "Gene Expression Quantification", 
                          workflow.type = "STAR - Counts")
saveRDS(query_tcga_gc, "data/query_TCGA_STAD.rds")
GDCdownload(query = query_tcga_gc, directory = "data/")
TCGA_STAD_se <- GDCprepare(query = query_tcga_gc, directory = "data/")
```

```{r}
# using days_to_last_follow_up and days_to_death to filtering
# after filtering 443 samples retained

ids <- !is.na(TCGA_STAD_se$days_to_last_follow_up) | !is.na(TCGA_STAD_se$days_to_death)
ids %>% table

TCGA_STAD_se <- TCGA_STAD_se[, ids]
saveRDS(TCGA_STAD_se, "data/TCGA_STAD_se.rds")
```

# mitophyage related genes (MRGs)
```{r}
library(msigdbr)
library(KEGGREST)

msigdbr_collections()

df <- msigdbr(species = "Homo sapiens", subcategory = "CP:REACTOME")
MRGs <- df %>% dplyr::filter(grepl("MITOPHAGY", gs_name, ignore.case = TRUE)) %>% dplyr::pull(gene_symbol) %>% unique
saveRDS(MRGs, "data/MRGs.rds")
```

```{r}
hsa_pathways <- keggList(database = "pathway")
```


# analysis
## single cell
```{r, message=FALSE, warning=FALSE}
# start a new session
library(magrittr)
library(DropletUtils)
library(scran)
library(scater)
library(bluster)
library(batchelor)
library(BiocParallel)
library(RColorBrewer)
library(dplyr)
library(tidyr)
library(ggplot2)
library(ggpubr)
library(fs)
library(stringr)
library(stringr.plus)
library(scDblFinder)
```

```{r}
samples <- dir_ls("data/sc/unfiltered/", type = "directory")

# patients selected
patients <- paste(c(5846, 5866, 5931, 6207, 6342, 6592, 6709), "t", sep = "_")
samples <- samples[str_replace(basename(samples), "[0-9]{1,}$", "") %in% patients]

sces <- read10xCounts(samples = samples, 
                      sample.names = basename(samples), 
                      col.names = T, 
                      type = "sparse", 
                      row.names = "symbol",
                      compressed = TRUE)
sces %>% colData
sces %>% rowData
sces

# add sample information
sces$patients <- str_extract_before(sces$Sample, "_", which = "first")
sces$patients %>% table

sces$condition <- ifelse(substr(sces$Sample, 6, 6) == "n", "normal", 
                         ifelse(substr(sces$Sample, 6, 6) == "t", "tumor", 
                                ifelse(substr(sces$Sample, 6, 6) == "p", "pbmc", NA)))
sces$condition %>% table

sces$replicate <- str_extract(string = sces$Sample, pattern = "[0-9]{1,}$")
sces$replicate %>% table

set.seed(101)
em <- emptyDrops(m = sces, test.ambient = TRUE)
saveRDS(em, "data/sc/emptyDrops_res.rds")
# xx <- removeAmbience(y = sces, ambient = metadata(em)$ambient, groups = rep(1, ncol(sces)))

# remove empty drops and low expression genes
sces <- sces[, which(em$FDR < 0.001)]
# sces <- sces[rownames(metadata(em)$ambient), ]
```

```{r}
# qc
is_mito <- grepl("^mt-", rownames(sces), ignore.case = T)
is_mito %>% table
rownames(sces)[is_mito]

sces <- addPerCellQCMetrics(sces, subsets = list(mito = is_mito))

sces$qc_mito <- isOutlier(sces$subsets_mito_percent, log = T, type = "both")
sces$qc_mito %>% table

sces$qc_sum <- isOutlier(sces$sum, log = T, type = "lower", nmads = 2)
sces$qc_sum %>% table

sces$qc_detected <- isOutlier(sces$detected, log = T, type = "lower", nmads = 2)
sces$qc_detected %>% table

sces %>% 
  colData %>% 
  as.data.frame %>% 
  ggplot(aes(sum, detected)) +
  geom_point(aes(color = qc_mito), size = 0.5, alpha = 1/3) +
  scale_x_continuous() +
  scale_y_continuous()

ids <- sces$qc_sum | sces$qc_detected | sces$qc_mito
ids %>% table
sces <- sces[, !ids]

# qc - feature
stats <- perFeatureQCMetrics(sces)
qc <- stats$detected > 0.01
qc %>% table
sces <- sces[qc, ]

saveRDS(sces, "data/sces.rds")
```

```{r}
# start a new session
sces <- readRDS("data/sces.rds")

set.seed(101)
qcluster <- quickCluster(x = sces, min.size = 50, block = sces$Sample)
qcluster %>% table

sces <- computePooledFactors(x = sces, clusters = qcluster)
sces <- logNormCounts(sces)
# sces <- multiBatchNorm(sces, batch = sces$Sample)
var <- modelGeneVar(sces, block = sces$Sample)
var[order(var$bio, decreasing = T), ]
hvgs <- getTopHVGs(var, var.threshold = 0)
hvgs %>% str
```

```{r}
set.seed(101)
sces <- correctExperiments(sces, 
                           batch = sces$Sample, 
                           subset.row = hvgs, 
                           correct.all = TRUE, 
                           PARAM = FastMnnParam(auto.merge = TRUE))
```











































