---
title: "Seurat"
author: "yincy"
date: "3/5/2020"
output:
    html_document:
        toc: true
        toc_depth: 3
        toc_float: true 
        number_section: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE}
library(Seurat)
library(magrittr)
```

```{r, echo=FALSE}
load(file = "/home/yincy/git/Data/Seurat/cluster-tutorial/pbmc.filter.RData")
```


# Object Overview  
The Seurat object is a class allowing for the storage and manipulation of single-cell data. Previous version of the Seurat object were designed primarily with scRNA-seq data in mind. However, with the development of new technologies allowing for multiple modes of data to be collected from the same set of cells, we have redesigned the Seurat 3.0 object to allow for greater flexibility to work with all these data types in a cohesive framework.  

**At the top level, the Seurat object serves as a collection of Assay and DimReduc objects, representing expression data and dimensionality reductions of the expression data, respectively**. The Assay objects are designed to hold expression data of a single type, such as RNA-seq gene expression, CITE-seq ADTs, cell hashtags, or imputed gene values. DimReduc objects represent transformations of the data contained within the Assay object(s) via various dimensional reduction techniques such as PCA. For class-specific details, including more in depth description of the slots, please see the wiki sections for each class.  

## Seurat  
### Slots  
**Slot**                | **Function**  
------------------------|--------------------------
`assays`                | A list of assays within this object
`meta.data`             | Cell-level meta data
`active.assay`          | Name of active, or default, assay
`active.ident`          | Identity classes for the current object
`graphs`                | A list of nearest neighbors graphs
`reductions`            | A list of DimReduc objects
`project.name`          | User-defined project name (optional)
`tools`                 | Empty list. Tool developer can store any internal data from their methods here
`misc`                  | Empty slot. User can store additional information here
`version`               | Seurat version used when creating the object


### Object Information  
Summary information about `Seurat` objects can be had quickly and easily using standard R functions.   

Object shape/dimensions can be found using the `dim`, `ncol`, and `nrow` functions; cell and feature names can be found using the `colnames` and `rownames` functions, respectively, or the `dimnames` function.   
```{r}
rownames(pbmc) %>% head()
colnames(pbmc) %>% head()
dim(pbmc)
```


A vector of names of `Assay`, `DimReduc`, and `Graph` objects contained in a Seurat object can be had by using names.  
```{r}
names(pbmc)
```


### Data Access  
Pulling specific `Assay`, `DimReduc`, or `Graph` objects can be done with the double `[[` extract operator.   

Adding new objects to a `Seurat` object is also done with the double `[[` extract operator; Seurat will figure out where in the Seurat object a new associated object belongs.  
```{r}
pbmc[['RNA']]
pbmc[["tsne"]]
```


Accessing data from an `Seurat` object is done with the `GetAssayData` function. Adding expression data to either the `counts`, `data`, or `scale.data` slots can be done with `SetAssayData.`   

New data must have the same cells in the same order as the current expression data. Data added to counts or data must have the same features as the current expression data.  
```{r}
GetAssayData(object = pbmc, slot = "data") %>% .[1:5, 1:5]
```

Cell-level meta data can be accessed with the single `[` extract operator or using the $ sigil.   

Pulling with the `$` sigil means only one bit of meta data can be pulled at a time, though tab-autocompletion has been enabled for it, making it ideal for interactive use.   

Adding cell-level meta data can be set using the single `[` extract operator as well, or by using `AddMetaData`.  

cell-level meta data is stored as a data frame  
```{r}
colnames(x = pbmc[[]])
```

```{r}
pbmc[[]] %>% head()
```

```{r}
pbmc[[c("nFeature_RNA", "nCount_RNA")]] %>% head()
```

```{r}
pbmc$percent.mt %>% head()
```

The `HVFInfo` (get `H`ighly `V`ariable `F`eature `Info`rmation) function pulls feature mean and dispersion from an `Assay` object. The vector of variable features can be pulled with the `VariableFeatures` function. `VariableFeatures` can also set the vector of variable features.    
```{r}
HVFInfo(object = pbmc) %>% head()
```


```{r}
VariableFeatures(object = pbmc) %>% head()
```

A vector of standard deviations for `DimReduc` stored within the Seurat object can be found with `Stdev`.  
```{r}
Stdev(pbmc, reduction = "pca")
```


### Methods  
Methods for the `Seurat` class can be found within the following:  
```{r}
Seurat_methods <- utils::methods(class = "Seurat") %>% as.character()
data.frame(methods = gsub(pattern = "[.Seurat]", "", Seurat_methods) %>% unique())
```



## Assays  
The `Assay` class stores single cell data.  

For typical scRNA-seq experiments, a Seurat object will have a single Assay ("RNA"). This assay will also store multiple 'transformations' of the data, including raw counts (@counts slot), normalized data (@data slot), and scaled data for dimensional reduction (@scale.data slot).  

For more complex experiments, an object could contain multiple assays. These could include multi-modal data types (CITE-seq antibody-derived tags, ADTs), or imputed/batch-corrected measurements. Each of those assays has the option to store the same data transformations as well.  

### Slots  

**Slot**           | **Function**
-------------------|------------------------------------
`counts`           | Stores unnormalized data such as raw counts or TPMs 
`data`             | Normalized data matrix  
`scale.data`       | Scaled data matrix  
`key`              | A character string to facilitate looking up features from a specific `Assay`  
`var.features`     | A vector of features identified as variable  
`meta.features`    | Feature-level meta data  

### Object Information  
Summary information about `Assay` objects can be had quickly and easily using standard R functions. Object shape/dimensions can be found using the `dim`, `ncol`, and `nrow` functions; cell and feature names can be found using the `colnames` and `rownames` functions, respectively, or the `dimnames` function.   

```{r}
rna <- pbmc@assays$RNA
rna
```

```{r}
dim(rna)
```

```{r}
rownames(rna) %>% head()
```

```{r}
colnames(rna) %>% head()
```

### Data Access  
Accessing data from an `Assay` object is done in several ways. Expression data is accessed with the `GetAssayData` function. Pulling expression data from the data slot can also be done with the single `[` extract operator. Adding expression data to either the `counts`, `data`, or `scale.data` slots can be done with `SetAssayData`. New data must have the same cells in the same order as the current expression data.  

```{r}
rna[1:5, 1:5]
```

```{r}
GetAssayData(object = rna, slot = "scale.data") %>% .[1:5, 1:5]
```


Feature-level meta data can be accessed with the double `[[` extract operator. Adding feature-level meta data can be set using the double `[[` extract operator as well.   

The `HVFInfo` function serves a specific version of the double `[[` extract operator, pulling certain columns from the meta data.  

```{r}
colnames(rna[[]])
```

```{r}
rna[[]] %>% head()
```



```{r}
HVFInfo(object = rna, selection.method = "vst") %>% head()
```

```{r}
rna[[c("vst.mean", "vst.variance")]] %>% head()
```

```{r}
VariableFeatures(rna) %>% head()
```

The key  
```{r}
Key(rna)
Key(rna) <- "myRNA"
```

Pull a feature from the RNA assay on the Seurat level  
```{r}
FetchData(object = pbmc, vars = "ident") %>% head()
```


### Mehtods  
Methods for the `Assay` class can be found with the following:  
```{r}
utils::methods(class = "Assay")
```


## DimReduc  
The `DimReduc` object represents a dimentional reduction taken upon the Seurat object.  

### Slots  

**Slot**                    | **Function** 
----------------------------|----------------------
`cell.embeddings`           | A matrix with cell embeddings  
`feature.loadings`          | A matrix with feature loadings  
`feature.loadings.projected`| A matrix with feature projected feature loadings  
`assay.used`                | Assay used to calculate this dimensional reduction  
`stdev`                     | Standard deviation for the dimensional reduction  
`key`                       | A character string to facilitate looking up features from a specific `DimReduc` 
`jackstraw`                 | Results from the `JackStraw` function  
`misc`                      | ...


### Object Information  
Summary information about `DimReduc` objects can be had quickly and easily using standard R functions. The number of cell embeddings and feature loadings can be found with `ncol` and `nrow`, respectively, or `dim` for both. The number of dimensions calculated can be found with `length`; feature and cell names can be found with `rownames` and `colnames`, respectively, or the `dimnames` function.   
```{r}
pca <- pbmc@reductions$pca
pca
```


```{r}
dim(pca)
```

```{r}
length(pca)
```

```{r}
rownames(pca) %>% head()
```

```{r}
colnames(pca) %>% head()
```


### Data Access  
Accessing data from an `DimReduc` object is done in several ways. **Feature loadings** is accessed with the `Loadings` function. Pulling feature loadings can also be done with the single `[` extract operator. Use of the single `[` extract operator will pull from the `feature.loadings.projected` slot unless it's not set, in which case it will pull from `feature.loadings`; the Loadings function can be forced to pull from one slot or the other.  

```{r}
pca[1:5, 1:5] # rowname is feature
```

```{r}
Loadings(object = pca, projected = F) %>% .[1:5, 1:5] # rowname is feature
```

**Cell embeddings** can be accessed with either the `Embeddings` function or the double `[[` extract operator.  
```{r}
pca[[1:5, 1:5]] # rowname is cell
```

```{r}
Embeddings(object = pca)[1:5, 1:5]  # rowname is cell 
```

Other data accessors include `DefaultAssay` for the name of the `Assay` used to calculate this dimensional reduction, `Key` for the key of the `DimReduc`, and `Stdev` for the vector of standard deviations.  

The Key can be used to pull cell embeddings for specific dimensions from the Seurat level  
```{r}
Key(object = pca)
```

```{r}
DefaultAssay(pca)
```

### Methods  
```{r}
utils::methods(class = "DimReduc")
```


# Seurat v3.0 Command List  
## Standard Workflow  
The standard Seurat workflow takes raw single-cell expression data and aims to find clusters within the data. For full details, please read our tutorial. This process consists of data normalization and variable feature selection, data scaling, a PCA on variable features, construction of a shared-nearest-neighbors graph, and clustering using a modularity optimizer. Finally, we use a t-SNE to visualize our clusters in a two-dimensional space.  
```{r}
pbmc.counts <- Read10X(data.dir = "/home/yincy/git/Data/Seurat/cluster-tutorial/filtered_gene_bc_matrices/hg19/")
pbmc <- CreateSeuratObject(counts = pbmc.counts)
pbmc <- NormalizeData(object = pbmc)
pbmc <- FindVariableFeatures(object = pbmc)
pbmc <- ScaleData(object = pbmc)
pbmc <- RunPCA(object = pbmc)
pbmc <- FindNeighbors(object = pbmc)
pbmc <- FindClusters(object = pbmc)
pbmc <- RunTSNE(object = pbmc)
DimPlot(object = pbmc, reduction = "tsne")
```


## Seurat Object Integration  
With Seurat v3.0, weâ€™ve made improvements to the Seurat object, and added new methods for user interaction. We also introduce simple functions for common tasks, like subsetting and merging, that mirror standard R functions.  
```{r}
# Get cell and feature names, and total numbers
colnames(pbmc)
Cells(pbmc)
rownames(pbmc)
ncol(pbmc)
nrow(pbmc)
```

```{r}
# Get cell identity classes
Idents(object = pbmc)
levels(x = pbmc)

# Stash cell identity classes
pbmc[["old.ident"]] <- Idents(object = pbmc)
pbmc <- StashIdent(object = pbmc, save.name = "old.ident")

# Set identity classes
Idents(object = pbmc) <- "CD4 T cells"
Idents(object = pbmc, cells = 1:10) <- "CD4 T cells"

# Set identity classes to an existing column in meta data
Idents(object = pbmc, cells = 1:10) <- "orig.ident"
Idents(object = pbmc) <- "orig.ident"

# Rename identity classes
pbmc <- RenameIdents(object = pbmc, `CD4 T cells` = "T Helper cells")
```

```{r}
# Subset Seurat object based on identity class, also see ?SubsetData
subset(x = pbmc, idents = "B cells")
subset(x = pbmc, idents = c("CD4 T cells", "CD8 T cells"), invert = TRUE)

# Subset on the expression level of a gene/feature
subset(x = pbmc, subset = MS4A1 > 3)

# Subset on a combination of criteria
subset(x = pbmc, subset = MS4A1 > 3 & PC1 > 5)
subset(x = pbmc, subset = MS4A1 > 3, idents = "B cells")

# Subset on a value in the object meta data
subset(x = pbmc, subset = orig.ident == "Replicate1")

# Downsample the number of cells per identity class
subset(x = pbmc, downsample = 100)
```

```{r}
# Merge two Seurat objects
merge(x = pbmc1, y = pbmc2)
# Merge more than two Seurat objects
merge(x = pbmc1, y = list(pbmc2, pbmc3))
```


## Data Access  
Accessing data in Seurat is simple, using clearly defined accessors and setters to quickly find the data needed.  
```{r}
# View metadata data frame, stored in object@meta.data
pbmc[[]]

# Retrieve specific values from the metadata
pbmc$nCount_RNA
pbmc[[c("percent.mito", "nFeature_RNA")]]

# Add metadata, see ?AddMetaData
random_group_labels <- sample(x = c("g1", "g2"), size = ncol(x = pbmc), replace = TRUE)
pbmc$groups <- random_group_labels
```

```{r}
# Retrieve or set data in an expression matrix ('counts', 'data', and 'scale.data')
GetAssayData(object = pbmc, slot = "counts")
pbmc <- SetAssayData(object = pbmc, slot = "scale.data", new.data = new.data)
```

```{r}
# Retrieve or set data in an expression matrix ('counts', 'data', and 'scale.data')
GetAssayData(object = pbmc, slot = "counts")
pbmc <- SetAssayData(object = pbmc, slot = "scale.data", new.data = new.data)
```

```{r}
# Get cell embeddings and feature loadings
Embeddings(object = pbmc, reduction = "pca")
Loadings(object = pbmc, reduction = "pca")
Loadings(object = pbmc, reduction = "pca", projected = TRUE)
```

```{r}
# FetchData can pull anything from expression matrices, cell embeddings, or metadata
FetchData(object = pbmc, vars = c("PC_1", "percent.mito", "MS4A1"))
```


## Visualization in Seurat v3.0  
Seurat has a vast, ggplot2-based plotting library. All plotting functions will return a ggplot2 plot by default, allowing easy customization with ggplot2.  
```{r}
# Dimensional reduction plot for PCA or tSNE
DimPlot(object = pbmc, reduction = "tsne")
DimPlot(object = pbmc, reduction = "pca")

# Dimensional reduction plot, with cells colored by a quantitative feature
FeaturePlot(object = pbmc, features = "MS4A1")

# Scatter plot across single cells, replaces GenePlot
FeatureScatter(object = pbmc, feature1 = "MS4A1", feature2 = "PC_1")
FeatureScatter(object = pbmc, feature1 = "MS4A1", feature2 = "CD3D")

# Scatter plot across individual features, repleaces CellPlot
CellScatter(object = pbmc, cell1 = "AGTCTACTAGGGTG", cell2 = "CACAGATGGTTTCT")

VariableFeaturePlot(object = pbmc)

# Violin and Ridge plots
VlnPlot(object = pbmc, features = c("LYZ", "CCL5", "IL32"))
RidgePlot(object = pbmc, feature = c("LYZ", "CCL5", "IL32"))

# Heatmaps
DoHeatmap(object = pbmc, features = heatmap_markers)
DimHeatmap(object = pbmc, reduction = "pca", cells = 200)

# New things to try!  Note that plotting functions now return ggplot2 objects, so you can add themes, titles, and options
# onto them
VlnPlot(object = pbmc, features = "MS4A1", split.by = "groups")
DotPlot(object = pbmc, features = c("LYZ", "CCL5", "IL32"), split.by = "groups")
FeaturePlot(object = pbmc, features = c("MS4A1", "CD79A"), blend = TRUE)
DimPlot(object = pbmc) + DarkTheme()
DimPlot(object = pbmc) + labs(title = "2,700 PBMCs clustered using Seurat and viewed\non a two-dimensional tSNE")
```

Seurat provides many prebuilt themes that can be added to ggplot2 plots for quick customization  

**Theme**                | **Function**
-------------------------|------------------------------
`DarkTheme`              | Set a black background with white text 
`FontSize`               | Set font size for various elements of a plot  
`NoAxes`                 | Remove axes and axis text
`NoLegend`               | Remove all legend elements  
`RestoreLegend`          | Restores a legend after removal  
`RotatedAxis`            | Rotates x-axis labels  

```{r}
# Plotting helper functions work with ggplot2-based scatter plots, such as DimPlot, FeaturePlot, CellScatter, and
# FeatureScatter
plot <- DimPlot(object = pbmc) + NoLegend()

# HoverLocator replaces the former `do.hover` argument It can also show extra data throught the `information` argument,
# designed to work smoothly with FetchData
HoverLocator(plot = plot, information = FetchData(object = pbmc, vars = c("ident", "PC_1", "nFeature_RNA")))

# FeatureLocator replaces the former `do.identify`
select.cells <- FeatureLocator(plot = plot)

# Label points on a ggplot object
LabelPoints(plot = plot, points = TopCells(object = pbmc[["pca"]]), repel = TRUE)
```


## Multi-Assay Features  
With Seurat v3.0, you can easily switch between different assays at the single cell level (such as ADT counts from CITE-seq, or integrated/batch-corrected data). Most functions now take an assay parameter, but you can set a Default Assay to aviod repetitive statements.  

```{r}
cbmc <- CreateSeuratObject(counts = cbmc.rna)
# Add ADT data
cbmc[["ADT"]] <- CreateAssayObject(counts = cbmc.adt)
# Run analyses by specifying the assay to use
NormalizeData(object = cbmc, assay = "RNA")
NormalizeData(object = cbmc, assay = "ADT", method = "CLR")

# Retrieve and set the default assay
DefaultAssay(object = cbmc)
DefaultAssay(object = cbmc) <- "ADT"
DefaultAssay(object = cbmc)

# Pull feature expression from both assays by using keys
FetchData(object = cbmc, vars = c("rna_CD3E", "adt_CD3"))

# Plot data from multiple assays using keys
FeatureScatter(object = cbmc, feature1 = "rna_CD3E", feature2 = "adt_CD3")
```

