---
title: "RNAseq-analysis-in-R"
author: "yincy"
date: "1/5/2020"
output: 
    html_document:
        toc: true
        toc_depth: 2
        highlight: tango
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```


## Introduction  
[Tutorial from this site](http://combine-australia.github.io/RNAseq-R/)  

This tutorial is about bulk RNA-seq data analysis, data is from this paper PMID: [ 25730472](https://www.ncbi.nlm.nih.gov/pubmed/25730472)  

This study examines the expression profiles of basal stem-cell enriched cells (B) and committed luminal cells (L) in mammary gland of virgin, pregnant and lactating mice. Six group are present, with one for each combination of cell type and mouse status. Each group contains two biological replications.  


## Reading data into R
```{r, echo = FALSE}
library(edgeR)
library(limma)
library(gplots)
library(org.Mm.eg.db)
library(RColorBrewer)
library(Glimma)
library(magrittr)
library(tidyverse)
```


```{r}
seqdata <- read.table(file = "data/Mouse-mammary/GSE60450_Lactation-GenewiseCounts.txt",
                      stringsAsFactors = F,
                      sep = "\t",
                      header = T)
```

```{r}
sampleinfo <- read.table(file = "data/Mouse-mammary/SampleInfo_Corrected.txt", 
                         sep = "\t",
                         header = T,
                         stringsAsFactors = F)
```


## Format the data  
```{r}
countdata <- seqdata[, -c(1, 2)]
rownames(countdata) <- seqdata[, 1]
```


```{r}
colnames(countdata) <- substr(x = colnames(countdata), start = 1, stop = 7)
head(countdata)
```

```{r}
identical(colnames(countdata), sampleinfo$SampleName)
```


## Filtering to remove lowly expressed genes  
remove genes that are expressed at a counts-per-million (CPM) below 0.5 in at least two samples.  
```{r}
mcpm <- cpm(countdata)
keep <- rowSums(mcpm) >= 2
table(keep)
```

```{r}
count_keep <- countdata[keep, ]
```

A CPM of 0.5 is used as it corresponds to a count of 10-15 for the library sizes in this data set. If the count is any smaller, it is considered to be very low, indicating that the associated gene is not expressed in that sample.   

A reqiurement for expression in two or more libraries is ued as each group contains two replicates. This ensures that a gene will be retained if it is only expressed in one group. Smaller CPM thresholds are usually appropriate for large libraries. As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.5.    

You should filter with CPMs rather than filtering on the counts directly, as the latter does not account for difference in library size between samples.  
```{r}
plot(mcpm[,1], countdata[, 1], ylim = c(0,50), xlim = c(0,3))
abline(h = 10, v = 0.5, col = "blue")
```


## Convert counts to DGEList object  
```{r}
y <- DGEList(counts = count_keep)
names(y)
y$counts %>% head()
y$samples
```



## Quality Control  
### Library Size and distribution plots  
```{r}
barplot(y$samples$lib.size, names.arg = colnames(y), las = 2)
title(main = "Barplot of Library Sizes")
```

```{r}
logcounts <- cpm(y, log = TRUE)
boxplot(logcounts, xlab = "", ylab = "Log2 counts per million", las = 2)
abline(h = median(logcounts), col = "blue")
title(main = "Boxplots of logCPMs")
```


### Multidimensional scaling plots  
An MDSplot is a visulisation of a principle components analysis, which determines the greatest sources of variation in the data.  

```{r}
unique(sampleinfo$Status)
```

```{r}
col_status <- brewer.pal(n = 3, name = "Set1")[as.factor(sampleinfo$Status)]
```

```{r}
plotMDS(y, col = col_status, pch = c(1, 4)[as.factor(sampleinfo$CellType)])
legend("topleft", fill = brewer.pal(n = 3, name = "Set1"), legend = unique(sampleinfo$Status), bty = "o")
legend("topright", legend = as.factor(unique(sampleinfo$CellType)), pch = c(1,4), bty = "o")
title("Status & Cell Types")
```


### Hierarchical clustering with heatmaps  
```{r}
var_genes_500 <- apply(logcounts, 1, var) %>% sort(decreasing = T) %>% head(n = 500) %>% names()
cols <- colorRampPalette(colors = c("red", "white", "blue"))

col_cell <- c("purple", "orange")[as.factor(sampleinfo$CellType)]

heatmap.2(logcounts[var_genes_500, ], 
          col = rev(cols(50)),
          main = "Top 500 most variable genes",
          ColSideColors = col_cell, 
          scale = "row",
          trace = "none")
```


## Normalisation for composition bias  
TMM normalization is performed to eliminate composition biases between libraries. This generates a set of normalization factors, where the product of these factors and the library size defines the effective library size.   

The `calcNormFactors` function calculates the normalization factors between libraries. TMM normalisation (and most scaling normalisation methods) scale relative to one sample.  

```{r}
y <- calcNormFactors(y)
y$samples
```

The normalization factors multiply to unity across all libraries. A normalization factor below one indicates that the library size will be scaled down, as there is more suppression (i.e., composition bias) in that library relative to the other libraries. This is also equivalent to scaling the counts upwards in that sample. Conversely, a factor above one scales up the library size and is equivalent to downscaling the counts.  


The last two samples seems abnormal  
```{r}
par(mfrow = c(2,2))

# unnormalized data
plotMD(logcounts, column = 7)
abline(h = 0, col = "grey")
plotMD(logcounts, column = 11)
abline(h = 0, col = "grey")

# TMM normalized data
plotMD(y, column = 7)
abline(h = 0, col = "grey")
plotMD(y, column = 11)
abline(h = 0, col = "grey")
```

## Differential expression with limma-voom  
### Creating the design matrix  
First we need to create a design matrix for the groups. There are many different ways to set up your design matrix, and it is dictated by what comparisions you would like to test.   
```{r}
group <- paste(sampleinfo$CellType, sampleinfo$Status, sep = "_")
design <- model.matrix(~ 0 + group)
colnames(design) <- gsub("group", "", colnames(design))
```


### Voom transform the data  
Once we have our design matrix ready to go, we can perform our voom transformation. Voom will automatically adjust the library sizes using the `norm.factors` already calculated.   

The voom transformation uses the experiment design matrix, and produces an `EList` object. We can add `plot=TRUE` to generate a plot of the mean-variance trend. This plot can also tell us if there are any genes that look really variable in our data, and if we’ve filtered the low counts adequately.   

```{r}
v <- voom(y, design = design, plot = T)
```

Compare lo2CPM before and after lib.size normalization  
```{r}
par(mfrow = c(1, 2))
boxplot(logcounts, xlab = NULL, ylab = "Logs counts per million", las = 2, main = "Unnormalised logCPM")
abline(h = median(logcounts), col = "blue")

boxplot(v$E, xlab = NULL, ylab = "Log counts per million", las = 2, main = "Voom transformed logCPM")
abline(h = median(v$E), col = "blue")
```


### Testing for differential expression  
```{r}
fit <- lmFit(v)
names(fit)
```

`lmFit` estimates group means according to the design matrix, as well as gene-wise variances.  

Since we are interested in differences between groups, we need to specify which comparisons we want to test. The comparison of interest can be specified using the makeContrasts function.  

ere, we are interested in knowing which genes are differentially expressed between the pregnant and lactating group in the basal cells. This is done by defining the null hypothesis as basal.pregnant - basal.lactate = 0 for each gene.   

>Note that the group names must exactly match the column names of the design matrix.  

```{r}
colnames(design)
cont_matrix <- makeContrasts(B.PregVsLac = basal.pregnant - basal.lactate, levels = design)
```

Now we can apply the contrasts matrix to the `fit` object to get the statistics and estimated parameters of our comparision that we are interested in.  
```{r}
fit_cont <- contrasts.fit(fit = fit, contrasts = cont_matrix)
```

The final step is to call the `eBayes` function, which performs empirical Bays shrinkage on the variances, and estimates moderated t-statistics and the associated p-values.  
```{r}
fit_cont <- eBayes(fit = fit_cont)
```

Check the dimensions of the fit object  
```{r}
dim(fit_cont)
```


We can use the *limma* `decideTests` function to generate a quick summary of DE genes for the contrasts.  
```{r}
summa.fit <- decideTests(object = fit_cont)
summary(summa.fit)
```


The *limma* `topTable` function summarises the output in a table format. Significant DE genes for a particular comparison can be identified by selecting genes with a p-value smaller than a chosen cut-off value and/or a fold change greater than a chosen value in this table.  

By default the table will be sorted by the **B statistic**, which is the **log-odds of differential expression**. Usually the B statistic and p-value ranking will be the same, but this is not always the case.  

The `topTable` command will always output the top 10 genes by default, even if they are not statistically significant. We can specify the coefficient we are interested in by the name we used in the contrast matrix (“B.PregVsLac”), or by the column number.  
```{r}
topTable(fit = fit_cont, coef = "B.PregVsLac", sort.by = "p")
```

same as the above  
```{r}
topTable(fit = fit_cont, coef = 1, sort.by = "p")
```


### Adding annotation and saving the results  
```{r}
columns(org.Mm.eg.db)
```

```{r}
ann <- AnnotationDbi::select(x = org.Mm.eg.db, 
              keys = rownames(fit_cont),
              keytype = "ENTREZID",
              columns = c("ENTREZID", "SYMBOL", "GENENAME"))
```

```{r}
table(ann$ENTREZID == rownames(fit_cont))
```

```{r}
fit_cont$genes <- ann
```

Now when we run the `topTable` command, the annotation information should be includeed in the output.  
```{r}
topTable(fit_cont, coef = 1, sort.by = "p", number = 20)
```

To get the full table we can specify `n = "Inf"`  
```{r}
limma_res <- topTable(fit = fit_cont, coef = 1, sort.by = "p", number = "Inf")
```


**A note about deciding how many genes are significant**: In order to decide which genes are differentially expressed, we usually talk about a cut-off of 0.05 on the adjusted p-value, NOT the raw p-value.  This is because we are testing more than 15000 genes, and the chances of finding differentially expressed genes is very high when you do that many tests. Hence we need to control the false discovery rate, which is the adjusted p-value column in the results table. What this means is that if 100 genes are significant at a 5% false discovery rate, we are willing to accept that 5 will be false positives. Note that the `decideTests` function displays significant genes at 5% FDR.    


### Plots after testing for DE  
```{r}
par(mfrow = c(1,2))
plotMD(object = fit_cont, coef = 1, status = summa.fit[, "B.PregVsLac"], values = c(-1, 1))

volcanoplot(fit = fit_cont, coef = 1, highlight = 20, names = fit_cont$genes$SYMBOL)
```

## Gene Set Testing  
Sometimes there is quite a long list of differentially expressed genes to interpret after a differential expression analysis, and it is usually infeasible to go through the list one gene at a time trying to understand it’s biological function. A common downstream procedure is gene set testing, which aims to understand which pathways/gene networks the differentially expressed genes are implicated in.  

There are a number of different ways to go about testing for enrichment of biological pathways, and the test you choose usually depends on the question you’re asking.  

There are two kinds of tests: self-contained and competitive gene set tests.  
1. **Self-contained tests**: which include the `ROAST` procedure, ask the question "Are the genes in the set/pathway differentially expressed as a whole?"  
2. **Competitive gene set tests**: like `goana` and `camera` ask the question whether the differentially expressed genes tend to be over-represented in the gene set, compared to all the other genes in the experiment.  


### Gene ontology testing with goana  
First, we will perform a gene ontology (GO) enrichment analysis using the goana function in limma. There are approximately 20,000 GO terms, and they are split into three categories: BP (biological process), MF (molecular function) and CC (cellular component). goana uses annotation from the appropriate Bioconductor package and can be used for any of the five species specified (Hs, Mm, Rn, Dm or Pt). goana has an advantage over other methods, such as DAVID, in that there is the option to take into account the gene length bias inherent in RNA-Seq data.  

Suppose we want to identify GO terms that are over-represented in the basal lactating group compared to the basal pregnancy group. This can be achieved by applying the goana function to the differential expression results of that comparison. goana takes the fit.cont object, the coefficient of interest and the species. The top set of most enriched GO terms can be viewed with the topGO function.  

```{r}
go <- limma::goana(de = fit_cont, coef = "B.PregVsLac", species = "Mm")
topGO(go, n = 10)
```


First, we will perform a gene ontology (GO) enrichment analysis using the goana function in limma. There are approximately 20,000 GO terms, and they are split into three categories: BP (biological process), MF (molecular function) and CC (cellular component). goana uses annotation from the appropriate Bioconductor package and can be used for any of the five species specified (Hs, Mm, Rn, Dm or Pt).  

goana has an advantage over other methods, such as DAVID, in that there is the option to take into account the gene length bias inherent in RNA-Seq data.  

Suppose we want to identify GO terms that are over-represented in the basal lactating group compared to the basal pregnancy group. This can be achieved by applying the goana function to the differential expression results of that comparison. goana takes the fit.cont object, the coefficient of interest and the species.  


```{r}
m <- match(rownames(fit_cont), seqdata$EntrezGeneID)
gene_length <- seqdata$Length[m]
go_length <- goana(fit_cont, coef = "B.PregVsLac", species = "Mm", covariate = gene_length)
topGO(go_length)
```











