---
title: "edgeR"
author: "yincy"
date: "11/27/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```

```{r load required packages}
library(edgeR)
library(GEOquery)
library(magrittr)
```


## Overview  
- classic `edgeR`: multigroup experiment  
- glm `edgeR`: multifactor experiment of any complexity    
    + likelihood ratio tests  
    + quasi-likelihood F-tests  

A particular feature of edgeR functionality, both classic and glm, are empirical Bayes methods that permit the estimation of gene-specific biological variation, even for experiments with minimal levels of biological replication.  


edgeR can be applied to differential expression at the gene, exon, transcript or tag level. In fact, read counts can be summarized by any genomic feature. edgeR analyses at the exon level are easily extended to detect differential splicing or isoform-specific differential expression.  


The **quasi-likelihood method** is highly recommended for differential expression analyses of bulk RNA-seq data as it gives stricter error rate control by accounting for the uncertainty in dispersion estimation.  


The **likelihood ratio test** can be useful in some special cases such as single cell RNA-seq and datasets with no replicates.   


## Quick start  
```{r, eval=FALSE}
x <- read.delim(file = "TableOfCounts.txt", row.names = "Symbol")
group <- factor(x = c(1,1,2,2))
y <- DGEList(counts = x, group = group)
y <- calcNormFactors(object = y)
design <- model.matrix(~0+group)
y <- estimateDisp(y, design = design)
```


Perform quasi-likelihood F-tests  
```{r, eval=FALSE}
fit <- glmQLFit(y, design = design)
qlf <- glmQLFTest(glmfit = fit, coef = 2)
topTags(qlf)
```


Perform likelihood ratio tests  
```{r, eval=FALSE}
fit <- glmFit(y,design = design)
lrt <- glmLRT(glmfit = fit, coef = 2)
topTags(lrt)
```


## Overview of capabilities  
edgeR performs differential abundance analysis for pre-defined genomic features. Although not strictly necessary, *it usually desirable that these genomic features are non-overlapping.*  

We usually count reads that overlap any exon for the given gene, including the UTR as part of the first exon.   

GDEList components:  
    - `counts`: numeric matrix of read counts  
    - `samples`: containing information about the samples or libraries  
    - `genes`: containing information about the genes or genomic features  
    - `group`: vector or factor giving the experimental group/condition for each sample/library  
    - `lib.size`: for the libarry size or sequencing depth for each sample. If not specified it will be computed from the column sums of the counts.  


## Filtering  
Genes with very low counts across all libraries provide little evidence for differential expression.   

In the biological point of view, a gene must be expressed at some minimal level before it is likely to be translated into a protein or to be biologically important.   

In addition, the pronounced discreteness of these counts interferes with some of the statistical approximations that are used later in the pipeline. These genes should be filtered out prior to further analysis.  

As a rule of thumb, genes are dropped if they canâ€™t possibly be expressed in all the samples for any of the conditions.  

Usually a gene is required to have a count of 5-10 in a library to be considered expressed in that library.  

Users should also **filter with count-per-million (CPM)** rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.  

Filtering out low expression genes by `filterByExpr()` function.  

The function accesses the group factor contained in y in order to compute the minimum group size, but the filtering is performed independently of which sample belongs to which group so that no bias is introduced.  

The group factor or the experimental design matrix can also be given directly to the `filterByExpr` function if not already set in the DGEList object.  


## Normalization  
Normalization is only necessary for sample-specific effects.  

The `calcNormFactors` function normalizes for RNA composition by finding a set of scaling factors for the library sizes that minimize the log-fold changes between the samples for most genes.  

The default method for computing these scale factors uses a trimmed mean of Mvalues (TMM) between each pair of samples. We call the product of the *original library size* and the scaling factor the *effective library size*. The *effective library size* replaces the original library size in all downsteam analyses.    

TMM is recommended for most RNA-Seq data where the majority (more than half) of the genes are believed not differentially expressed between any pair of the samples.   

The normalization factors of all the libraries multiply to unity. A normalization factor below one indicates that a small number of high count genes are monopolizing the sequencing, causing the counts for other genes to be lower than would be usual given the library size. As a result, the library size will be scaled down, analogous to scaling the counts upwards in that library. Conversely, a factor above one scales up the library size, analogous to downscaling the counts.  


# Chapter 4 Case studies  
## RNA-seq of oral carcinomas vs matched normal tissue  













