---
title: "scater"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction  
`scater` provides tools for visualization of single-cell transcriptomic data. It is based on the `SingleCellExperiment` class (from the `SingleCellExperiment` package), and thus is interoperable with many other Bioconductor packages such as `scran`, `scuttle` and `iSEE`. To demonstrate the use of the various `scater` functions, we will load in the classic Zeisel dataset:  

```{r}
library(scRNAseq)
sce.zeisel <- readRDS("../workflow/Single-Cell/scRNA-seq-online/data/sce.zeisel.rds")
```

# Diagnostic plots for quality control  
Quality control remove damaged cells and poorly sequenced libaries is a common step in single-cell analysis pipelines. We will use some utilities from the `scuttle` package (conveniently loaded for us when we load `scater`) to compute the ususal quality control metrics for this dataset.  

```{r}
library(scater)

sce.zeisel <- addPerCellQC(sce.zeisel, 
                           subsets = list(Mito = grep("mt-", rownames(sce.zeisel))))
```

Metadata variables can be plotted against each other using the `plotColData()` function, as shown below. We expect to see an increasing number of detected genes with increasing ttoal count. Each point represents a cell that is coloured according to its tissue of origin.  

```{r}
colData(sce.zeisel) %>% colnames()

plotColData(sce.zeisel, x = "sum", y = "detected", colour_by = "tissue")
```

Here, we have plotted the total count for each cell against the mitochondrial content. Well-behaved cells should have a large number of expressed features and a low percentage of expression from feature controls. High percentage expression from feature controls and few expressed features are indicative of blank and failed cells. For some varity, we have faceted by the tissue of origin.  

```{r}
plotColData(sce.zeisel, 
            x = "sum", 
            y = "subsets_Mito_percent", 
            other_fields = "tissue") +
  facet_wrap(~ tissue)
```

On the gene level, we can look at a plot that shows the top 50 (by default) most-expressed features. Each row in the plot below corresponds to a gene; each bar corresponds to the expression of a gene in a single cell; and the circle indicates the median expression of each gene, with which genes are sorted. We expect to see the “usual suspects”, i.e., mitochondrial genes, actin, ribosomal protein, MALAT1. A few spike-in transcripts may also be present here, though if all of the spike-ins are in the top 50, it suggests that too much spike-in RNA was added. A large number of pseudo-genes or predicted genes may indicate problems with alignment.  

```{r}
plotHighestExprs(sce.zeisel, exprs_values = "counts")
```

Variable-level metrics are computed by the  `getVarianceExplained()` function (after normalization, see below). This calculates the percentage of variance of each gene’s expression that is explained by each variable in the `colData` of the  `SingleCellExperiment` object. We can then use this to determine which experimental factors are contributing most to the variance in expression. This is useful for diagnosing batch effects or to quickly verify that a treatment has an effect.  

```{r}
# Computing variance explained on the log-counts, so that the statistics reflect changed in relative expression
sce.zeisel <- scran::computeSumFactors(sce.zeisel, cluster = sce.zeisel$level1class)
sce.zeisel <- logNormCounts(sce.zeisel)

vars <- getVarianceExplained(x = sce.zeisel, 
                             variables = c("tissue", "total mRNA mol", "sex", "age"), 
                             exprs_values = "logcounts")

vars %>% head
```

```{r}
plotExplanatoryVariables(vars)
```

# Visualizing expression values  
The `plotExpression()` function makes it easy to plot expression values for subset of genes or features. This can be particularly useful for further examination of features identified from differential expression testing, pseudotime analysis or other analyses. By default, it uses expression values in the  `"logcounts"` assay, but this can be changed through the  `exprs_values` argument.  

```{r}
plotExpression(object = sce.zeisel, 
               features = rownames(sce.zeisel)[1:6], 
               x = "level1class") +
  theme(axis.text.x = element_text(angle = -40, hjust = 0)) +
  scale_x_discrete(name = NULL)
```

Setting `x` will determine the covariate to be shown on the x-axis. This can be a field in the column metadata or the name of a feature (to obtain the expression profile across cells). Categorical covariates will yield grouped violins as shown above, with one panel per feature. By comparison, continuous covariates will generate a scatter plot in each panel, as shown below.   

```{r}
plotExpression(sce.zeisel, 
               features = rownames(sce.zeisel)[1:6], 
               x = rownames(sce.zeisel)[10])
```

The points can also be coloured, shaped or resized by the column metadata or expression values.  

```{r}
plotExpression(sce.zeisel, 
               features = rownames(sce.zeisel)[1:6], 
               x = "level1class", 
               colour_by = "tissue") +
  theme(axis.text.x = element_text(angle = -40, hjust = 0))
```

Directly plotting the gene expression without any x or other visual parameters will generate a set of grouped violin plots, coloured in an aesthetically pleasing manner.  

```{r}
plotExpression(sce.zeisel, 
               features = rownames(sce.zeisel)[1:6])
```

# Dimensionality reduction  
## Principal components analysis
Principal components analysis (PCA) is often performed to denoise and compact the data prior to downstream analyses. The  `runPCA()` function provides a simple wrapper around the base machinery in `BiocSingular` for computing PCs from log-transformed expression values. This stores the output in the  `reducedDims` slot of the `SingleCellExperiment`, which can be easily retrieved (along with the percentage of variance explained by each PC) as shown below:  

```{r}
sce.zeisel <- runPCA(x = sce.zeisel)
reducedDim(sce.zeisel, "PCA") %>% str
```

By default, `runPCA()` uses the top 500 genes with the highest variances to compute the first PCs. This can be tuned by specifying `subset_row` to pass in an explicit set of genes of interest, and by using `ncomponents` to determine the number of components to compute. The `name` argument can also be used to change the name of the result in the reducedDims slot.  

```{r}
sce.zeisel <- runPCA(x = sce.zeisel, 
                     name = "PCA2", 
                     subset_row = rownames(sce.zeisel)[1:1000], 
                     ncomponents = 25)

reducedDim(sce.zeisel, "PCA2") %>% str
```

## Other dimensionality reduction methods  
t-distributed stochastic neighbour embedding (t-SNE) is widely used for visualizing complex single-cell data sets. The same procedure described for PCA plots can be applied to generate t-SNE plots using `plotTSNE`, with coordinates obtained using  `runTSNE` via the `Rtsne` package. We strongly recommend generating plots with different random seeds and perplexity values, to ensure that any conclusions are robust to different visualizations.  

```{r}
# Perplexity of 10 just chosen here arbitrarily
set.seed(1000)
sce.zeisel <- runTSNE(sce.zeisel, perplexity = 10)
reducedDim(sce.zeisel, "TSNE") %>% str
```

A more common pattern involves using the pre-existing PCA results as input into the t-SNE algorithm. This is useful as it improves speed by using a low-rank approximation of the expression matrix; and reduces random noise, by focusing on the major factors of variation. The code below uses the first 10 dimensions of the previously computed PCA result to perform the t-SNE.  

```{r}
set.seed(1000)
sce.zeisel <- runTSNE(x = sce.zeisel, 
                      perplexity = 50, 
                      dimred = "PCA", 
                      n_dimred = 10)

reducedDim(sce.zeisel, "TSNE") 
```

The same can be done for uniform manifold with approximate projection (UMAP) via the `runUMAP()` function, itself based on the `uwot` package.  

```{r}
sce.zeisel <- runUMAP(x = sce.zeisel)
reducedDim(sce.zeisel, "TSNE")
```

## Visualizing reduced dimensions  
Any dimensionality reduction result can be plotted using the  `plotReducedDim` function. Here, each point represents a cell and is coloured according to its cell type label.  

```{r}
plotReducedDim(sce.zeisel, 
               dimred = "PCA", 
               colour_by = "level1class")
```

```{r}
plotTSNE(object = sce.zeisel, 
         colour_by = "Snap25")
```

```{r}
plotPCA(sce.zeisel, 
        colour_by = "Mog")
```

Multiple components can be plotted in a series of pairwise plots. When more than two components are plotted, the diagonal boxes in the scatter plot matrix show the density for each component.  

```{r}
sce.zeisel <- runPCA(x = sce.zeisel, ncomponents = 20)
plotPCA(object = sce.zeisel, 
        ncomponents = 4, 
        colour_by = "level1class")
```

We separate the execution of these functions from the plotting to enable the same coordinates to be re-used across multiple plots. This avoids repeatedly recomputing those coordinates just to change an aesthetic across plots.  

# Utilities for custom visualization  
We provide some helper functions to easily convert from a  `SingleCellExperiment` object to a `data.frame` for use in, say, `ggplot2` functions. This allows users to create highly customized plots that are not covered by the existing `scater` functions.  

The `ggcells()` function will intelligently retrieve fields from: 

1. `colData()`  
2. `assays()`  
3. `altExps()`  
4. `reducedDims()` 

to create a single `data.frame` for immediate use. In the example below, we create boxplots of Snap25 expression stratified by cell type and tissue of origin:  

```{r}
ggcells(x = sce.zeisel, mapping = aes(x = level1class, y = Snap25)) +
  geom_boxplot() +
  facet_wrap(~tissue)
```

Reduced dimension results are easily pulled in to create customized equivalents of the `plotReducedDim()` output. In this example, we create a t-SNE plot faceted by tissue and coloured by Snap25 expression:  

```{r}
ggcells(sce.zeisel, mapping = aes(x = TSNE.1, y = TSNE.2, color = Snap25)) +
  geom_point() +
  stat_density_2d() +
  facet_wrap(~tissue) +
  scale_color_distiller(direction = 1)
```

It is also straightforward to examine the relationship between the size factors on the normalized gene expression:  

```{r}
ggcells(sce.zeisel, mapping = aes(x = sizeFactor, y = Actb)) +
  geom_point() +
  geom_smooth()
```

`ggfeatures()` function, whcih retrive values from:  

1. `rowData()`  
2. columns of `assays()`

```{r}
colnames(sce.zeisel) <- colnames(sce.zeisel) %>% make.names()

ggfeatures(x = sce.zeisel, mapping = aes(x = featureType, y = X1772062111_E06)) +
  geom_violin()
```



