---
title: "DropletUtils"
author: "yincy"
date: "7/16/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

Droplet-based single-cell RNA sequencing (scRNA-seq) technologies allow researchers to obtain transcriptome-wide expression profiles for thousands of cells at once. Briefly, each cell is encapsulated in a droplet in a oil-water emulsion, along with a bead containing reverse transcription primers with a unique barcode sequence. After reverse transcription inside the droplet, each cell’s cDNA is labelled with that barcode (referred to a “cell barcode”). Bursting of the droplets yields a pool of cDNA for library preparation and sequencing. Debarcoding of the sequences can then be performed to obtain the expression profile for each cell.

This package implements some general utilities for handling these data after quantification of expression. In particular, we focus on the 10X Genomics platform, providing functions to load in the matrix of unique molecule identifier (UMI) counts as well as the raw molecule information. Functions are also available for downsampling the UMI count matrix or the raw reads; for distinguishing cells from empty droplets, based on the UMI counts; and to eliminate the effects of barcode swapping on Illumina 4000 sequencing machine.  

# Reading in 10X Genomics data
## From the UMI count matrix
The `CellRanger` pipeline from 10X Genomics will process the raw sequencing data and produce a matrix of UMI counts. Each row of this matrix corresponds to a gene, while each column corresponds to a cell barcode. This is saved in a single directory for each sample, usually named like <OUTPUT>/outs/filtered_gene_bc_matrices/<GENOME>1. We mock up an example directory below using some simulated data:  

```{r}
# to generate the file
example(write10xCounts, echo = F)
dir.name <- tmpdir
list.files(dir.name)
```
The `matrix.mtx` file contains the UMI counts, while the other two files contain the cell barcodes and the gene annotation. We can load this into memory using the `read10xCounts()` function, which returns a `SingleCellExperiment` object containing all of the relevant information. This includes the barcode sequence for each cell (column), as well as the identifier and symbol for each gene (row).  

```{r}
sce <- read10xCounts(samples = dir.name)
```

The counts themselves are loaded as a sparse matrix, specifically a `dgCMatrix` from the `Matrix` package. This reduces memory usage by only storing the non-zero counts, which is useful for sparse scRNA-seq data with lots of dropouts.  

```{r}
class(counts(sce))
```

Users can also load multiple samples at once by supplying a character vector to `read10xCounts()`. This will return a single `SingleCellExperiment` where all of the individual matrices are combined by column. Obviously, this only makes sense when the same set of genes is being used across samples.  

## From the molecule information file
`CellRanger` will also produce a molecule information file (molecule_info.h5) that contains… well, information about the transcript molecules. This includes the UMI sequence, the cell barcode sequence, the gene to which it was assigned, and the number of reads covering the molecule. For demonstration purposes, we create an example molecule information file below:

```{r}
set.seed(1000)
mol.info.file <- ""
```


## Downsampling on the reads
Given multiple batches of very different sequencing depths, it can be beneficial to downsample the deepest batches to match the coverage of the shallowest batches. This avoids differences in technical noise that can drive clustering by batch. The scuttle package provides some utilities to downsample count matrices, but technically speaking, downsampling on the reads is more appropriate as it recapitulates the effect of differences in sequencing depth per cell. This can be achieved by applying the `downsampleReads` function to the molecule information file containing the read counts:  

```{r}
set.seed(100)
no.sampling <- downsampleReads(mol.info.file, prop = 1)
sum(on.sampling)
```

```{r}
with.sampling <- downsampleReads(mol.info.file, prop = 0.5)
sum(with.sapling)
```

The above code will downsample the reads to 50% of the original coverage across the experiment. However, the function will return a matrix of UMI counts, so the final total count may not actually decrease if the libraries are sequenced to to saturation! Users should use  `downsampleMatrix()` instead if they want to guarantee similar total counts after downsampling.


## Computing barcode ranks
A useful diagnostic for droplet-based data is the barcode rank plot, which shows the (log-)total UMI count for each barcode on the y-axis and the (log-)rank on the x-axis. This is effectively a transposed empirical cumulative density plot with log-transformed axes. It is useful as it allows users to examine the distribution of total counts across barcodes, focusing on those with the largest counts. To demonstrate, let us mock up a count matrix:  

```{r}
set.seed(0)
my.counts <- DropletUtils::
```



























