---
title: "03 Multiple Alignments"
author: "YinCY"
date: 2023-02-22
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Introduction
The `DNAMultipleAlignment` `RNAMultipleAlignment` and `AAMultipleAlignment` classes allow users to represent groups of aligned DNA, RNA or amino acid sequences as a single object. **The frame of reference for aligned sequences is static, so manipulation of these objects is confined to be non-destructive**. In practice, this means that these objects contain slots to mask ranges of rows and columns on the original sequence. These masks are then respected by methods that manipulate and display the objects, allowing the user to remove or expose columns and rows without invalidating the original alignment.


# Creation and masking
To create a MultipleAlignment, call the appropriate read function to read in and parse the original alignment. There are functions to read clustaW, Phylip and Stolkholm data formats.  

```{r}
library(Biostrings)
library(magrittr)

origMAlign <- readDNAMultipleAlignment(filepath = system.file("extdata", 
                                                              "msx2_mRNA.aln", 
                                                              package = "Biostrings"), 
                                       format = "clustal")

phylipMAlign <- readAAMultipleAlignment(filepath = system.file("extdata", 
                                                               "Phylip.txt", 
                                                               package = "Biostrings"), 
                                        format = "phylip")
rownames(origMAlign) <- c("Human", "Chimp", "Cow", "Mouse", "Rat", "Dog", "Chicken", "Salmon")
origMAlign
```

To see a more detailed version of your `MultipleAlignment` object, you can use the `detail` method, which will show the details of the alignment interleaved and without the rows and columns that you have masked out.  

```{r}
origMAlign %>% detail()
```


Applying masks is a simple matter of specifying which ranges to hide.  

```{r}
maskTest <- origMAlign
rowmask(maskTest) <- IRanges(start = 1, end = 3)
maskTest %>% rowmask()
```

```{r}
colmask(maskTest) <- IRanges(start = c(1, 1000), end = c(500, 2343))
maskTest %>% colmask()
maskTest
```


remove row and column masks by assigning NULL  

```{r}
rowmask(maskTest) <- NULL
colmask(maskTest) <- NULL
maskTest
```

When setting a mask, you might want to specify the rows or columns to keep, rather than to hide. To do that, use the `invert` argument.   

```{r}
rowmask(maskTest, invert = TRUE) <- IRanges(start = 4, end = 8) 
maskTest %>% rowmask()
maskTest
```

```{r}
colmask(maskTest, invert = TRUE) <- IRanges(start = 501, end = 999)
maskTest %>% colmask()
maskTest
```

In addition to being able to invert these masks, you can also choose the way in which the ranges you provide will be merged with any existing masks. The append argument allows you to specify the way in which new mask ranges will interact with any existing masks. By default, these masks will be the "union" of the new mask and any existing masks, but you can also specify that these masks be the mask that results from when you "intersect" the current mask and the new mask, or that the new mask simply "replace" the current mask. The append argument can be used in combination with the invert argument to make things even more interesting. In this case, the inversion of the mask will happen before it is combined with the existing mask.  

```{r}
rowmask(maskTest) <- NULL
colmask(maskTest) <- NULL

rowmask(maskTest) <- IRanges(start = 1, end = 3)
maskTest
```

```{r}
rowmask(maskTest, append = "intersect") <- IRanges(start = 2, end = 5)
maskTest
```

```{r}
rowmask(maskTest, append = "replace") <- IRanges(start = 5, end = 8)
maskTest
```

```{r}
rowmask(maskTest, append = "replace", invert = TRUE) <- IRanges(start = 5, end = 8)
maskTest
```

```{r}
rowmask(maskTest, append = "union") <- IRanges(start = 7, end = 8)
maskTest
```

The function `maskMotif` works on `MultipleAlignment` objects too, and takes the same arguments that it does elsewhere. `maskMotif` is useful for masking occurances of a string from columns where it is present in the consensus sequence.  

```{r}
tataMasked <- maskMotif(origMAlign, motif = "TATA")
tataMasked %>% colmask()
```

`maksGaps` also operates om columns and will mask columns based on the fraction of each column that contains gap `min.fraction` along with the width of columns that contain this fraction of gaps `min.block.width`.  

```{r}
automasked <- maskGaps(origMAlign, min.fraction = 0.5, min.block.width = 4)
automasked
automasked %>% colmask()
```

Sometimes you may want to cast your `MultipleAlignment` to be a matrix for usage eslewhere. `as.matrix` is supported for these circumstances. The ability to convert one object into another is not very unusual so why mention it? Because when you cast your object, the masks **WILL** be considered so that the masked rows and columns will be left out of the matrix object.


```{r}
full <- as.matrix(origMAlign)
full %>% dim

partial <- as.matrix(automasked)
partial %>% dim
```


# Analytic utilities
Once you have masked the sequence, you can then ask questions about the properties of that sequence. For example, you can look at the alphabet frequency  of that sequence. The alphabet frequency will only be for the masked sequence.  

```{r}
alphabetFrequency(automasked, as.prob = TRUE)
```


You can also calculate a consensus matrix, extract the consensus string or look at the consensus views. These methods too will all consider the masking when you run them.  

```{r}
consensusMatrix(automasked, baseOnly = TRUE)[, 84:90]
```

```{r}
substr(consensusString(automasked), 80, 160)
```

```{r}
consensusViews(automasked)
```

You can also cluster the alignments based on their distance to each other. Because you must pass in a DNAStringSet, the clustering will also take into acount the masking. So for example, you can see how clustering the unmasked DNAMultipleAlignment will draw a funky looking tree.  

```{r}
sdist <- stringDist(as(origMAlign, "DNAStringSet"), method = "hamming")
clust <- hclust(d = sdist, method = "ward.D2")
plot(clust)
```

```{r}
groups <- cutree(clust, k = 2)
groups
```


# Exporting to file
One possible export option is to write to fasta files if you need to write your `MultipleAlignment` object object out as a fasta file, you can cast it to a DNAStringSet.  

```{r}
DNAStr <- as(origMAlign, "DNAStringSet")
writeXStringSet(DNAStr, "myFile.fa")
```

One other format that is of interest is the Phylip format. The Phylip format stores the column masking of your object as well as the sequence that you are exporting. So if you have masked the sequence and you write out a Phylip file, this mask will be recorded into the file you export. As with the fasta example above, any **rows** that you have masked out will be removed from the exported file. Any column mask will be marked in the output file.    

```{r}
phylipMAlign %>% colmask()
rowmask(phylipMAlign) <- IRanges(1, 1)
phylipMAlign
write.phylip(phylipMAlign, "phylip.txt")
```







