---
title: "Biostring"
author: "yincy"
date: "11/7/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load package, message=FALSE}
library(Biostrings)
library(BSgenome)
library(BSgenome.Hsapiens.UCSC.hg38)
library(magrittr)
```


## Overview  
The `Biostrings` package contains classes and functions for representing biological strings such as DNA, RNA and amino acids.   

In addition the package has functionality for pattern matching (short read alignment) as well as a pairwise alignment function implementing Smith-Waterman local alignments and Needleman-Wunsch global alignments used in classic sequence alignment (see (Durbin et al. 1998) for a description of these algorithms).   

There are also functions for reading and writing output such as FASTA files.  


## Representing Sequences  
There are two basic types of containers for representing strings. One container represents a single string (say a chromosome or a single short read) and the other container represnets a set of strings (say a set of short reads). There are different classes intended to represnet different types of sequences such as DNA or RNA sequences.  

```{r}
dna1 <- DNAString(x = "ACGT-N")
```

```{r, eval=FALSE}
DNAString(x = "ADE")
```


```{r}
dna2 <- DNAStringSet(x = c("ACGT", "GTCA", "GCTA"))
```

Note that the alphabet of a `DNAString` is an extended alphabet: - (for insertion) and N are allowed. In fact, `IUPAC` codes are allowed (these codes represent different characters, for example the code “M” represents either and “A” or a “C”).  

```{r IUPAC code}
IUPAC_CODE_MAP
```


```{r}
substr <- function(string, start, stop){
    x = subseq(x = string, start = start, end = stop)
    toString(x = x)
}

substr(dna1, 1, 3)
```


Indexing into a `DNAString` retrieves a subsequence (similar to the standard R function substr), whereas indexing into a `DNAStringSet` gives you a subset of sequences.  

```{r}
dna1[2:4] %>% toString()
```

```{r}
dna2[2:3]
```

Note that `[[` allows you to get a single element of a `DNAStringSet` as a `DNAString`. This is very similar to `[` and `[[` for lists.  

```{r}
dna2[[1]]
```


`DNAStringSet` objects can have names, like ordinary vectors  
```{r}
names(dna2) <- paste("seq", 1:3, sep = "")
dna2
```


The full set of string classes:  
- `DNAString[Set]`: DNA sequences.  
- `RNAString[Set]`: RNA sequences.  
- `AAString[Set]`: Amino acids sequences  
- `BString[Set]`: 'big' sequences, using any kind of letter.  

In addition you will often see references to `XString[Set]` in the documentation. An `XString[Set]` is basically any of the above classes.  

These classes seem very similar to standard `characters()` from base R, but there are important differences. The differences are mostly about efficiencies when you deal with either (a) many sequences or (b) very long strings (think whole chromosomes).  


## Basic functionality  
Basic character functionality is supported:  
- `length`, `names`.  
- `c` and `rev` (reverse the sequence).  
- `width`, `nchar` (number of characters in each sequence.  
- `==`, `duplicated`, `unique`.  
- `as.charcater` or `toString`: converts to a base character vector.  
- `sort`, `order`.  
- `chartr`: convert some letters into other letters.  
- `subseq`, `subseq<-`, `extractAt`, `replaceAt`.  
- `replaceLetterAt`.  

```{r}
width(dna2)
```


```{r}
sort(dna2)
```

```{r}
rev(dna2)
```

```{r}
dd2 <- DNAString("AGCCGGGCCATAAGGCA")
v4 <- Views(subject = dd2, start = 3:0, end = 5:8)
v4
```
```{r}
length(v4)
```

```{r}
v4[4:2]
```

```{r}
v12 <- Views(DNAString("TAATAATG"), start=-2:9, end=0:11)
v12 == DNAString(x = "TAA")
v12[v12 == DNAString(x = "TAA")]
```

```{r}
v12
width(v12)
start(v12)
end(v12)
```


## Biological functionality  
There are also functions which are related to the biological interpretation of the sequences, including  
- `reverse`: reverse the sequence  
- `complement`, `reverseComplement`: (reverse) complement the sequence  
- `translate`: translate the DNA or RNA sequence into amino acids  


```{r}
translate(dna2)
```

```{r}
dna1
reverseComplement(dna1)
complement(dna1)
```


## Counting letters  
We very often want to count sequences in various ways:  
- Compute the GC content of a set of sequences.  
- Compute the frequencies of dinucleotides in a set of sequences.  
- Compute a position weight matrix from a set of aligned sequences.  


There is a rich set of functions for doing this quickly.  
- `alphabetFrequency`, `letterFrequency`: Compute the frequency of all characters(`alphabetFrequency`) or only specific letters(`letterFrequency`).  

- `dinucleotideFrequency`, `trinucleotideFrequency`, `oligonucleotideFrequency`: compute frequencies of dinucleotides (2 bases), trinucleotides (3 bases) and oligonucleotides (general number of bases).  

- `letterFrequencyInSlidingView`: letter frequencies, but in sliding views along the string.  

- `consensusMatrix`: consensus matrix; almost a position weight matrix.  

```{r}
alphabetFrequency(dna1)
```


```{r}
alphabetFrequency(dna2)
```

```{r}
letterFrequency(x = dna2, letters = "GC")
```


```{r}
consensusMatrix(x = dna2, as.prob = TRUE)
```


# Matching  
## Pattern matching  
We often want to find patterns in (long) sequences. `Biostrings` have a number of functions for doing so  

- `matchPattern` and `vmatchPattern`: match a single sequence against one sequence (`matchPattern`) or more than one (`vmatchPattern`) sequences.  

- `matchPDict` and `vmatchPDict`: match a (possibly large) set of sequences against one sequence (`matchPDict`) or more than one (`vmatchPDict`) sequences.  

These functions allows a small set of mismatches and some small indels. The `Dict` term is used because the function builds a 'dictionary' over the sequences.  

There are also functions with similar naming using `count` instead of match (eg. `countPatterns`). These functions returns the number of matches instead of precise information about where the matches occur.  

In many ways, these functions are similar to using short read aligners like Bowtie. But these functions are designed to be **comprehensive** (return all matches satisfying certain criteria).  

```{r}
dnaseq <- DNAString(x = "GCCACCATGG")
views <- matchPattern(pattern = dnaseq, 
                      subject = BSgenome.Hsapiens.UCSC.hg38$chrY,
                      max.mismatch = 0)
views 
```
note how the return object of `matchPattern` looks like an `IRanges` but is really something called a Views.  

```{r}
kozak <- "GCCACCATGG"
for(i in seq_along(kozak)){
    BSgenome.Hsapiens.UCSC.hg38$chrY[start(kozak)[i]:(start(kozak)[i] + 10)] %>% print
}
```

```{r}
countPattern(pattern = "GCCACCATGG", 
             subject = BSgenome.Hsapiens.UCSC.hg38$chrY, 
             max.mismatch = 0)
```


```{r}
library(BSgenome.Scerevisiae.UCSC.sacCer2)
vmatchPattern(pattern = dnaseq, subject = Scerevisiae)
```

```{r}
vcountPattern(pattern = dnaseq, subject = Scerevisiae)
```

First, note how the return object of `vmatchPattern` is a `GRanges` given the exact information of where the string matches.  

Note **sequence we search for is its own reverse complement**, so we get hits on both strands (which makes sense). Obviously, not all sequences are like this.  
```{r}
dnaseq == reverseComplement(dnaseq)
```


## Specialized alignments  
There are a number of other, specialized, alignment functions in `Biostrings`.  

- `matchPWM`: a **position weight matrix** is a common way to represent for example a transcription factor binding motif (think sequence logos). This function allows you to search for such motifs in the genome.  

- `pairwiseAlignment`: This function implements pairwise alignments using dynamic programming; providing an interface to both the Smith-Waterman local alignment problem and the Needleman-Wunsch global alignment problems.  

- `trimLRpattern` (trim left-right pattern): Takes a set of sequences and looks for whether they start or end with a given (other sequence), for example a sequencing adapter. Used for trimming reads based on adapter sequences.  


`pairwiseAlignment` allows you to do pairwise alignments of millions of short reads against a single sequence, for example a gene or a transposable element. Few people use these algorithms for short read data, because the algorithms scale badly with the length of the sequence (ie. the genome), but they work fine for millions of reads as long as the reference sequence is short. In my opinion this approach might be very fruitful if you are particular interested in high-quality alignments to a specific small gene or region of the genome.  



# From vignette  
## The XString class and its subsetting operator [  
The `XString` is in fact a virtual class and therefore cannot be instanciated. Only subclasses (or subtypes) `BString`, `DNAString`, `RNAString` and `AAString` can. These classes are direct extensions of the XString class.  

```{r}
b <- BString(x = "I am a BString object")
```

```{r}
length(b)
```

```{r}
d <- DNAString(x = "TTGAAAA-CTC-N")
```

```{r}
length(d)
```

The differences with a BString object are  
- only letters from the **IUPAC** *extended genetic alphabet* and the gap letter (-) are allowed  
- each letter in the argument passed to the `DNAString` function is encoded in a special way before it's stored in the DNAString object.  


```{r}
d[3]
```

```{r}
d[7:12]
```

```{r}
d[]
```

```{r}
b[length(b):1]
```


In fact the subsetting operator for XString objects is not efficient and one should always
use the subseq method to extract a substring from a big string  

```{r}
bb <- subseq(x = b, start = 3, end = 6)
```

```{r}
dd1 <- subseq(x = d, end = 7)
dd2 <- subseq(x = d, start = 8)
```


to dump an XString object as a character vector (of length 1), use the `toString` method  
```{r}
toString(dd2)
```
 
Note that length(dd2) is equivalent to nchar(toString(dd2)) but the latter would be very inefficient on a big DNAString object.  


## The == binary operator for XString objects  
```{r}
bb == "am a"
dd2 != DNAString("TG")
```

When the 2 sides of == don’t belong to the same class then the side belonging to the “lowest” class is first converted to an object belonging to the class of the other side (the “highest” class).  

The class (pseudo-)order is character < BString < DNAString.  

When comparing an RNAString object with a DNAString object, U and T are considered equals  
```{r}
r <- RNAString(x = d)
```

```{r}
r == d
```


## The XString Views class and its subsetting operators [ and [[  
The Views virtual class is a general container for storing a set of views on an arbitrary Vector object, called the "subject".  

An `XStringViews` object contains a set of views on the same `XString` object called the subject string.  
```{r}
v4 <- Views(dd2, start = 3:0, end = 5:8)
```

```{r}
length(v4)
```

**Note that the last 2 views are out of limits**.  
You can select a subset of views from an `XStringViews` object:  
```{r}
v4[4:2]
```

```{r}
v4[[2]]
```

You can't extract a view that is out of limits.  

Note that, when start and end are numeric vectors and i is a single integer, **Views(b, start, end)[[i]]** is equivalent to **subseq(b, start[i], end[i])**.  

Subsetting also works with negative or logical values with the expected semantic (the same
as for R built-in vectors):  
```{r}
v4[-3]
```

```{r}
v4[c(TRUE, FALSE)]
```

Note that the logical vector is recycled to the length of v4.  


## A few more XStringViews objects  
```{r}
v12 <- Views(DNAString(x = "TAATAATG"), start = -2:9, end = 0:11)
```

```{r}
as(d, "Views")
v12[0]
v12[FALSE]
Views(d)
```


## The == binary operator for XStringViews objects  
This operator is the vectorized version of the == operator defined previously for XString objects  
```{r}
v12 == DNAString(x = "TAA")
```


```{r}
v12[v12 == v12[4]]
```

```{r}
v12[v12 == v12[1]]
```


## The start, end and width methods  
```{r}
start(v4)
```

```{r}
end(v4)
```

```{r}
width(v4)
```


# Pairwise Sequence Alignments  
## Introduction  
`pairwiseAlignment()` function aligns a set of *pattern* strings to a *subject* string in a global, local, or overlap (ends-free) fashion with or without affine gaps using either a fixed or quality-based substitution scoring scheme.  

This function's computation time is proportional to the product of the two string lengths being aligned.  

## Pairwise Sequence Alignment Problems  
The optimal pairwise sequence alignment is the pairwise sequence alignment with the largest score for the specified alignment type, substitution scoring scheme, and gap penalties.  

The pairwise sequence alignment types, substitution scoring schemes, and gap penalties influence alignment scores in the following manner.  

- **Pairwise Sequence Alignment Types**:  The type of pairwise sequence alignment determines the substring ranges to apply the substitution scoring and gap penalty schemes.  

For the three primary (global, local, overlap) and two derivative (subject overlap, pattern overlap) pairwise sequence alignment types, **the resulting substring ranges** are as follows.  
```{r}
knitr::include_graphics("alignment_type.PNG")
```
   

- **Substitution Scoring Schemes**:  The substitution scoring scheme sets the values for the aligned character pairings within the substring ranges determined by the type of pairwise sequence alignment. This scoring scheme can be fixed for character pairings or quality-dependent for character pairings.  

```{r}
pairwiseAlignment(pattern = c("succeed", "precede"), 
                  subject = "supersede")
```

The type of pairwise sequence alignment is set by specifying the *type* argument to be one of 'global', 'local', 'overlap', 'global-local', and 'local-global'.  

```{r}
pairwiseAlignment(pattern = c("succeed", "precede"), 
                  subject = "supersede", 
                  type = "local")
```

```{r}
pairwiseAlignment(pattern = c("succeed", "precede"), 
                  subject = "supersede", 
                  gapOpening = 0,
                  gapExtension = 1)
```

The substitution scoring scheme is set using three arguments, two of which are quality-based related (`patternQuality`, `subjectQuality`) and one is fixed substitution related (`substitutionMatrix` ). When the substitution scores are fixed by character pairing, the `substituionMatrix` argument takes a matrix with the appropriate alphabets as dimension names. The `nucleotideSubstitutionMatrix` function tranlates simple match and mismatch scores to the full spectrum of IUPAC nucleotide codes.  

```{r}
submat <- matrix(-1, nrow = 26, ncol = 26, dimnames = list(letters, letters))
diag(submat) <- 0
```

```{r}
pairwiseAlignment(pattern = c("succeed", "precede"), 
                  subject = "supersede", 
                  gapOpening = 0,
                  gapExtension = 1,
                  substitutionMatrix = submat,
                  type = "global")
```

When the substitution scores are quality-based, the `patternQuality` and `subjectQuality` arguments represent the equivalent of [x - 99] numeric quality values for the respective strings, and the optional `fuzzyMatrix` argument represents how the closely two characters match on a [0, 1] scale.  

The `patternQuality` and `subjectQuality` arguments accept quality measures in either a `PhredQuality`, `SolexaQuality`, or `IlluminaQuality` scaling.   

For `PhredQuality` and `IlluminaQuality` measures Q = [0, 99], the probability of an error in the base read is given by 10^-Q/10^ and for `SolexaQuality` measures Q = [-5, 99], they are given by 1-1/(1+10^-Q/10^).  

```{r}
pairwiseAlignment(pattern = c("succeed", "precede"),
                  subject = "supersede",
                  substitutionMatrix = submat,
                  gapOpening = 0,
                  gapExtension = 1,
                  scoreOnly = T)
```


```{r}
pairwiseAlignment(pattern = "syzygy",
                  subject = "zyzzyx",
                  type = "local")
```

```{r}
pairwiseAlignment(pattern = "syzygy",
                  subject = "zyzzyx",
                  type = "global")
```


```{r}
pairwiseAlignment(pattern = "syzygy",
                  subject = "zyzzyx",
                  type = "overlap")
```

```{r}
pairwiseAlignment(pattern = "syzygy",
                  subject = "zyzzyx",
                  type = "global-local")
```

```{r}
pairwiseAlignment(pattern = "syzygy",
                  subject = "zyzzyx",
                  type = "local-global")
```


## Pairwise Sequence Alignment Classes  
Biostrings package provide five specialty classes:  
    - `PairwiseAlignments`:  used to represent *elementwise alignments* between a **set of patterns** and a **set of subjects**.  
    
    - `PairwiseAlignmentsSingleSubject`: inherits from the *PairwiseAlignments* and their both the results of a fit from the `pairwiseAlignment` function, being used to represent all patterns aligning to a single subject.     
    
    - `PairwiseAlignmentsSingleSubjectSummary`:
    
    - `AlignedXStringSet`: hold the 'gapped' S~i~ substrings with a fix subsitution scoring scheme.  
    
    - `QualityAlignedXStringSet`: hold the gapped S~i~ substrings with a quality-based scoring scheme.      
    

```{r}
pa1 <- pairwiseAlignment(pattern = c("scceed", "precede"),
                         subject = "supersede")
class(pa1)
```

```{r}
summary(pa1)
```


```{r}
pa2 <- pairwiseAlignment(pattern = c("succeed", "precede"),
                         subject = "supersede",
                         substitutionMatrix = submat,
                         gapOpening = 0, 
                         gapExtension = 1)
class(pa2)
pattern(pa2) %>% class()
```


## Pairwise Sequence Alignment Helper Functions  
The `score`, `nedit`, `nmatch`, `nmismatch`, and `nchar` functions return numeric vectors containing information on the pairwise sequence alignment score, Levenshtein edit distance, number of matches, number of mismatches, and number of aligned characters respectively.  

```{r}
pa2 <- pairwiseAlignment(pattern = c("succeed", "precede"),
                         subject = "supersede",
                         gapOpening = 0,
                         gapExtension = 1,
                         substitutionMatrix = submat)
```

```{r}
score(pa2)
```

```{r}
nedit(pa2)
```

```{r}
nmatch(pa2)
```

```{r}
nmismatch(pa2)
```

```{r}
nchar(pa2)
```

```{r}
as.character(pa2)
```

```{r}
as.matrix(pa2)
```

```{r}
aligned(pa2)
```

```{r}
consensusMatrix(pa2, as.prob = T)
```

```{r}
summary(pa2)
```

```{r}
mismatchTable(pa2)
```

```{r}
mismatchSummary(pa2)
```


```{r}
pattern(pa2)
```

```{r}
subject(pa2)
```


## Edit Distances  
```{r}
pairwiseAlignment(pattern = "syzygy", 
                  subject = "zyzzyx")
```


```{r}
stringDist(x = c('a' = "zyzzyx", "b" = "syzygy", 'c' = "zyzzyx"), 
           ignoreCase = T, 
           diag = T, 
           upper = F)
```


```{r}
data("BLOSUM50")
BLOSUM50 %>% .[1:5, 1:5]
```

```{r}
nwdemo <- pairwiseAlignment(pattern = AAString(x = "PAWHEAE"), 
                            subject = AAString("HEAGAWGHEE"),
                            substitutionMatrix = BLOSUM50, 
                            gapOpening = 0, 
                            gapExtension = 8)

nwdemo
```

`compareStrings` combines two equal-length strings that are assumed to be aligned to be aligned into a single characater string containing that replaces mismatches with '?', insertions with '+', and deletions with '-'.  
```{r}
compareStrings(nwdemo)
```

```{r}
pid(nwdemo) # percent Sequence identity
```

```{r}
data("BLOSUM62")
pairwiseAlignment(pattern = AAString("PAWHEAE"), 
                  subject = AAString("HEAGAWGHEE"), 
                  substitutionMatrix = BLOSUM62, 
                  gapOpening = 12, 
                  gapExtension = 4)
```


## Removing Adapters from Sequence Reads  
```{r}
simulationReads <- function(N, adapter, experiment, substitutionRate = 0.01, gapRate = 0.001){
    chars <- strsplit(x = as.character(adapter), split = "")[[1]]
    sapply(seq_len(N), function(i, experiment, substitutionRate, gapRate){
        width <- experiment[["width"]][i]
        side <- experiment[["side"]][i]
        randomLetters <- function(n){
            sample(DNA_ALPHABET[1:4], n, replace = T)
        }
        randomLettersWithEmpty <- function(n){
            sample(c("", DNA_ALPHABET[1:4]), n, replace = T, prob = c(1 - gapRate, rep(gapRate/4, 4)))
        }
        nChars <- length(chars)
        
        value <- paste(ifelse(rbinom(nChars, 1, substitutionRate), randomLetters(nChars), chars), 
                       randomLettersWithEmpty(nChars), sep = "", collapse = "")
        
        if(side){
            value <- paste(c(randomLetters(36 - width), substring(value, 1, width)), sep = "", collapse = "")
        }else{
            value <- paste(c(substring(value, 37 - width, 36), randomLetters(36 - width)), sep = "", collapse = "")
        }
        
        value
    }, 
    experiment = experiment, substitutionRate = substitutionRate, gapRate = gapRate)
}

adapter <- DNAString("GATCGGAAGAGCTCGTATGCCGTCTTCTGCTTGAAA")

set.seed(123)
N <- 1000

experiemnt <- list(side = rbinom(N, 1, 0.5), width = sample(0:36, N, replace = T))
table(experiemnt[["side"]], experiemnt[["width"]])
```

```{r}
adapterStrings <- simulationReads(N, adapter, experiemnt, substitutionRate = 0.01, gapRate = 0.001)
adapterStrings <- DNAStringSet(adapterStrings)
```

```{r}
M <- 5000
randomStrings <- apply(matrix(sample(DNA_ALPHABET[1:4], 36*M, replace = T), 
                              nrow = M), 
                       MARGIN = 1, 
                       FUN = paste, 
                       collapse = "")

randomStrings <- DNAStringSet(x = randomStrings)
randomStrings
```

```{r}
submat1 <- nucleotideSubstitutionMatrix(match = 0, mismatch = -1, baseOnly = TRUE)
randomScores1 <- pairwiseAlignment(pattern = randomStrings, 
                                   subject = adapter, 
                                   substitutionMatrix = submat1, 
                                   gapOpening = 0, 
                                   gapExtension = 1, 
                                   scoreOnly = T)

randomScores1 %>% quantile(probs = seq(0.99, 1, 0.001))
```



```{r}
submat2 <- nucleotideSubstitutionMatrix(match = 1, 
                                        mismatch = -Inf, 
                                        baseOnly = TRUE, 
                                        type = "DNA")
randomScores2 <- pairwiseAlignment(pattern = randomStrings, 
                                   subject = adapter, 
                                   substitutionMatrix = submat2, 
                                   gapOpening = 0, 
                                   gapExtension = -Inf, 
                                   scoreOnly = TRUE, 
                                   type = "local")

randomScores2 %>% quantile(probs = seq(0.99, 1, 0.001))
```

```{r}
adapterAligns2 <- pairwiseAlignment(pattern = adapterStrings, 
                                    subject = adapter, 
                                    substitutionMatrix = submat2, 
                                    type = "local", 
                                    gapOpening = 0, 
                                    gapExtension = Inf)
table(score(adapterAligns2) > quantile(randomScores2, 0.99), experiemnt[["width"]])
```


# Multiple Alignments  
































