---
title: "TCGA"
author: "YinCY"
date: "9/15/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r, message=FALSE, warning=FALSE}
library(TCGAbiolinks)
library(dplyr)
library(DT)
```


# Searching GDC database
**Different sources: Legacy vs Harmonized**  

There are two available sources to download GDC data using TCGAbiolinks:  
- GDC Legacy Archive: provides access to an unmodified copy of data that was previously stored in CGHub and in the TCGA Data Portal hosted by the TCGA Data Coordinating Center (DCC), in which uses as references GRCh37(hg19) and GRCh36(hg18).  

- GDC harmonized database: data available was harmonized against GRCh38 (hg38) using GDC Bioinformatics Pipelines which provides methods to the standardization of biospecimen and clinical data.  


# Searching arguments
You can easily search GDC data using the `GDCquery` function.

|**Project**|**A list of valid project**|
|:--|:--|:--|
|data.category|A valid project. See `TCGAbiolinks::getDataCategorySummary(project)`||
|data.type|A data type to filter the files to download||
|workflow.type|GDC workflow type||
|legacy|Search in the legacy repository||
|access|Filter by access type. Possible values:controlled, open.||
|platform|Example:||
||CGH- 1x1M_G4447A|IlluminaGA_RNASeqV2|
||AgilentG4502A_07|IlluminaGA_mRNA_DGE|
||Human1MDuo|HumanMethylation450|
||HG-CGH-415K_G4124A|IlluminaGA_miRNASeq|
||HumanHap550|IlluminaHiSeq_miRNASeq|
||ABI|H-miRNA_8x15K|
||HG-CGH-244A|SOLiD_DNASeq|
||IlluminaDNAMethylation_OMA003_CPI|IlluminaGA_DNASeq_automated|
||IlluminaDNAMethylation_OMA002_CPI|HG-U133_Plus_2|
||HuEx- 1_0-st-v2|Mixed_DNASeq|
||H-miRNA_8x15Kv2|IlluminaGA_DNASeq_curated|
||MDA_RPPA_Core|IlluminaHiSeq_TotalRNASeqV2|
||HT_HG-U133A|IlluminaHiSeq_DNASeq_automated|
||diagnostic_images|microsat_i|
||IlluminaHiSeq_RNASeqSOLiD_DNASeq_curated|
||IlluminaHiSeq_DNASeqC|Mixed_DNASeq_curated|
||IlluminaGA_RNASeq|IlluminaGA_DNASeq_Cont_automated|
||IlluminaGA_DNASeq|IlluminaHiSeq_WGBS|
||pathology_reports|IlluminaHiSeq_DNASeq_Cont_automated|
||Genome_Wide_SNP_6|bio|
||tissue_images|Mixed_DNASeq_automated|
||HumanMethylation27|Mixed_DNASeq_Cont_curated|
||IlluminaHiSeq_RNASeqV2|Mixed_DNASeq_Cont|
|file.type|To be used in the legacy database for some platforms, to define which file types to be used.||
|barcode|A list of barcodes to filter the files to download|
|experimental.strategy|Filter to experimental strategy. Harmonized: WXS, RNA-Seq, miRNA-Seq, Genotyping Array. Legacy: WXS, RNA-Seq, miRNA-Seq, Genotyping Array, DNA-Seq, Methylation array, Protein expression array, WXS,CGH array, VALIDATION, Gene expression array,WGS, MSI-Mono-Dinucleotide Assay, miRNA expression array, Mixed strategies, AMPLICON, Exon array, Total RNA-Seq, Capillary sequencing, Bisulfite-Seq||
|sample.type|A sample type to filter the files to download||

## project options
```{r}
TCGAbiolinks::getGDCprojects()
```

## sample.type options
|**tissue.code**|**shortLetterCode**|**tissue.definition**|
|:--|:--|:--|
|01|TP|Primary Tumor|
|02|TR|Recurrent Tumor|
|03|TB|Primary Blood Derived Cancer - Peripheral Blood|
|04|TRBM|Recurrent Blood Derived Cancer - Bone Marrow|
|05|TAP|Additional - New Primary|
|06|TM|Metastatic|
|07|TAM|Additional Metastatic|
|08|THOC|Human Tumor Original Cells|
|09|TBM|Primary Blood Derived Cancer - Bone Marrow|
|10|NB|Blood Derived Normal|
|11|NT|Solid Tissue Normal|
|12|NBC|Buccal Cell Normal|
|13|NEBV|EBV Immortalized Normal|
|14|NBM|	Bone Marrow Normal|
|20|CELLC|Control Analyte|
|40|TRB|Recurrent Blood Derived Cancer - Peripheral Blood|
|50|CELL|	Cell Lines|
|60|XCL|	Cell Line Derived Xenograft Tissue|


# Examples
```{r}
query <- GDCquery(project = c("TCGA-GBM", "TCGA-LGG"), 
                  data.category = "DNA Methylation", 
                  legacy = F, 
                  platform = c("Illumina Human Methylation 450"), 
                  sample.type = "Recurrent Tumor")
```

```{r}
query %>% getResults() %>% datatable(filter = "top", 
                                     options = list(scrollX = T, keys = T, pageLength = 5), 
                                     rownames = F)
```

# Summary of available files per patient
```{r}
tab <- getSampleFilesSummary(project = "TCGA-ACC")
datatable(head(tab), 
          filter = "top", 
          options = list(scrollX = T, keys = T, pageLength = 5), 
          rownames = F)
```



















