---
title: "Chapter 12 Cell type annotation"
author: "YinCY"
date: "6/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Motivation
The most challenging task in scRNA-seq data analysis is arguably the interpretation of the results. Obtaining clusters of cells is fairly straightforward, but it is more difficult to determine what biological state is represented by each of those clusters. Doing so requires us to bridge the gap between the current dataset and prior biological knowledge, and the latter is not always available in a consistent and quantitative manner. Indeed, even the concept of a “cell type” is not clearly defined, with most practitioners possessing a “I’ll know it when I see it” intuition that is not amenable to computational analysis. As such, interpretation of scRNA-seq data is often manual and a common bottleneck in the analysis workflow.  

To expedite this step, we can use various computational approaches that exploit prior information to assign meaning to an uncharacterized scRNA-seq dataset. The most obvious sources of prior information are the curated gene sets associated with particular biological processes, e.g., from the Gene Ontology (GO) or the Kyoto Encyclopedia of Genes and Genomes (KEGG) collections. Alternatively, we can directly compare our expression profiles to published reference datasets where each sample or cell has already been annotated with its putative biological state by domain experts. Here, we will demonstrate both approaches with several different scRNA-seq datasets.  

## Assigning cell labels from reference data
## Overview
A conceptually straightforward annotation approach is to compare the single-cell expression profiles with previously annotated reference datasets. Labels can then be assigned to each cell in our uncharacterized test dataset based on the most similar reference sample(s), for some definition of “similar”. This is a standard classification challenge that can be tackled by standard machine learning techniques such as random forests and support vector machines. Any published and labelled RNA-seq dataset (bulk or single-cell) can be used as a reference, though its reliability depends greatly on the expertise of the original authors who assigned the labels in the first place.  

In this section, we will demonstrate the use of the SingleR method (Aran et al. 2019) for cell type annotation. This method assigns labels to cells based on the reference samples with the highest Spearman rank correlations, using only the marker genes between pairs of labels to focus on the relevant differences between cell types. It also performs a fine-tuning step for each cell where the correlations are recomputed with just the marker genes for the top-scoring labels. This aims to resolve any ambiguity between those labels by removing noise from irrelevant markers for other labels.  

### Using existing references
For demonstration purposes, we will use one of the 10X PBMC datasets as our test. While we have already applied quality control, normalization and clustering for this dataset, this is not strictly necessary. It is entirely possible to run `SingleR()` on the raw counts without any a priori quality control and filter on the annotation results at one’s leisure - see the book for an explanation.  

```{r}
sce.pbmc <- readRDS("f:/git/Data/Bioconductor/osca/sce.pbmc.rds")

# gene annotation
library(scater)
rownames(sce.pbmc) <- uniquifyFeatureNames(ID = rowData(sce.pbmc)$ID, 
                                           name = rowData(sce.pbmc)$Symbol)
library(EnsDb.Hsapiens.v86)
location <- mapIds(EnsDb.Hsapiens.v86, 
                   keys = rowData(sce.pbmc)$ID, 
                   keytype = "GENEID", 
                   column = "SEQNAME")

# cell detection
set.seed(100)
library(DropletUtils)
e.out <- emptyDrops(counts(sce.pbmc))
sce.pbmc <- sce.pbmc[, which(e.out$FDR <= 0.001)]

# quality control
stats <- perCellQCMetrics(sce.pbmc, subsets = list(Mito = which(location == "MT")))
high.mito <- isOutlier(stats$subsets_Mito_percent, type = "high")
sce.pbmc <- sce.pbmc[, !high.mito]

# normalization
library(scran)
set.seed(1000)

clusters <- quickCluster(sce.pbmc)
sce.pbmc <- computeSumFactors(sce.pbmc, cluster = clusters)
sce.pbmc <- logNormCounts(sce.pbmc)

# variance modeling
dec.pbmc <- modelGeneVarByPoisson(sce.pbmc)
top.pbmc <- getTopHVGs(dec.pbmc, prop = 0.1)

# dimensionality reduction
set.seed(10000)
sce.pbmc <- denoisePCA(sce.pbmc, subset.row = top.pbmc, technical = dec.pbmc)

set.seed(100000)
sce.pbmc <- runTSNE(sce.pbmc, dimred = "PCA")

set.seed(1000000)
sce.pbmc <- runUMAP(sce.pbmc, dimred = "PCA")

# clustering
g <- buildSNNGraph(sce.pbmc, k = 10, use.dimred = "PCA")
clust <- igraph::cluster_walktrap(g)$membership
colLabels(sce.pbmc) <- factor(clust)

sce.pbmc
```

The `celldex` contains a number of curated reference datasets, mostly assembled from bulk RNA-seq or microarray data of sorted cell types. These references are often good enough for most applications provided that they contain the cell types that are expected in the test population. Here, we will use a reference constructed from Blueprint and ENCODE data (Martens and Stunnenberg 2013; The ENCODE Project Consortium 2012); this is obtained by calling the `BlueprintEncode()` function to construct a `SummarizedExperiment` containing log-expression values with curated labels for each sample.  

```{r}
library(celldex)

ref <- BlueprintEncodeData()
colData(ref)
```

We call the `SingleR()` function to annotate each of our PBMCs with the main cell type labels from the Blueprint/ENCODE reference. This returns a `DataFrame` where each row corresponds to a cell in the test dataset and contains its label assignments. Alternatively, we could use the labels in ref$label.fine, which provide more resolution at the cost of speed and increased ambiguity in the assignments.  

```{r}
library(SingleR)

pred <- SingleR(test = sce.pbmc, 
                ref = ref, 
                labels = ref$label.main)

pred$labels %>% table
```

We inspect the results using a heatmap of the per-cell and label scores (Figure 12.1). Ideally, each cell should exhibit a high score in one label relative to all of the others, indicating that the assignment to that label was unambiguous. This is largely the case for monocytes and B cells, whereas we see more ambiguity between CD4+ and CD8+ T cells (and to a lesser extent, NK cells).  

```{r}
plotScoreHeatmap(results = pred)
```

We compare the assignments with the clustering results to determine the identity of each cluster. Here, several clusters are nested within the monocyte and B cell labels (Figure 12.2), indicating that the clustering represents finer subdivisions within the cell types. Interestingly, our clustering does not effectively distinguish between CD4+ and CD8+ T cell labels. This is probably due to the presence of other factors of heterogeneity within the T cell subpopulation (e.g., activation) that have a stronger influence on unsupervised methods than the a priori expected CD4+/CD8+ distinction.  

```{r}
tab <- table(Assigned = pred$pruned.labels, Cluster = colLabels(sce.pbmc))

# adding a pseudo-count of 10 to avoid strong color jump with just 1 cell
library(pheatmap)
library(viridis)
pheatmap(log2(tab + 1), color = viridis(100))
```

This episode highlights some of the differences between reference-based annotation and unsupervised clustering. The former explicitly focuses on aspects of the data that are known to be interesting, simplifying the process of biological interpretation. However, the cost is that the downstream analysis is restricted by the diversity and resolution of the available labels, a problem that is largely avoided by de novo identification of clusters. We suggest applying both strategies to examine the agreement (or lack thereof) between reference label and cluster assignments. Any inconsistencies are not necessarily problematic due to the conceptual differences between the two approaches; indeed, one could use those discrepancies as the basis for further investigation to discover novel factors of variation in the data.  

### Using custom references
We can also apply SingleR to single-cell reference datasets that are curated and supplied by the user. This is most obviously useful when we have an existing dataset that was previously (manually) annotated and we want to use that knowledge to annotate a new dataset in an automated manner. To illustrate, we will use the Muraro et al. (2016) human pancreas dataset as our reference.  

```{r}
library(scRNAseq)

sce.murao <- MuraroPancreasData()
# saveRDS(sce.murao, "F:/git/Data/Bioconductor/osca/sce.murao.rds")

# gene annotation
library(EnsDb.Hsapiens.v86)
library(stringr)

gene.symbol <- str_replace(rownames(sce.murao), "__chr[:alnum:]{1,2}$", "")
gene.id <- mapIds(EnsDb.Hsapiens.v86, 
                  keys = gene.symbol, 
                  keytype = "SYMBOL", 
                  column = "GENEID")

# remove duplicated genes or genes without ensembl ids
keep <- !is.na(gene.id) & !duplicated(gene.id)
sce.murao <- sce.murao[keep, ]
rownames(sce.murao) <- gene.id[keep]

# quality control
library(scater)
stats <- perCellQCMetrics(sce.murao)
qc <- quickPerCellQC(stats, 
                     percent_subsets = "altexps_ERCC_percent", 
                     batch = sce.murao$donor, 
                     subset = sce.murao$donor != "D28")

sce.murao <- sce.murao[, !qc$discard]

# normalization
library(scran)
set.seed(1000)
clusters <- quickCluster(sce.murao)
sce.murao <- computeSumFactors(sce.murao, clusters = clusters)
sce.murao <- logNormCounts(sce.murao)

sce.murao
```

```{r}
sce.murao <- sce.murao[, !is.na(sce.murao$label) & sce.murao$label != "nuclear"]
sce.murao$label %>% table
```

Our aim is to assign labels to our test dataset from Segerstolpe et al. (2016). We use the same call to `SingleR()` but with `de.method="wilcox"` to identify markers via pairwise Wilcoxon ranked sum tests between labels in the reference Muraro dataset. This re-uses the same machinery from Chapter 11; further options to fine-tune the test procedure can be passed via the `de.args` argument.  

```{r}
sce.seger <- SegerstolpePancreasData()
# saveRDS(sce.seger, "F:/git/Data/Bioconductor/osca/sce.seger.rds")

# gene annotation
library(EnsDb.Hsapiens.v86)
symbol <- rowData(sce.seger)$symbol
ens.id <- mapIds(EnsDb.Hsapiens.v86, keys = symbol, keytype = "SYMBOL", column = "GENEID")
ens.id <- ifelse(is.na(ens.id), symbol, ens.id)

# removing duplicated rows
keep <- !duplicated(ens.id)
sce.seger <- sce.seger[keep, ]
rownames(sce.seger) <- ens.id[keep]

# sample annotation
emtab.meta <- colData(sce.seger)[, c("cell type", "disease", "individual", "single cell well quality")]
colnames(emtab.meta) <- c("CellType", "Disease", "Donor", "Quality")
colData(sce.seger) <- emtab.meta

sce.seger$CellType <- gsub(" cell", "", sce.seger$CellType)
sce.seger$CellType <- paste(toupper(substr(sce.seger$CellType, 1, 1)), 
                            substring(sce.seger$CellType, 2), sep = "")

# quality control
low.qual <- sce.seger$Quality == "low quality cell"

library(scater)
stats <- perCellQCMetrics(sce.seger)
qc <- quickPerCellQC(stats, percent_subsets = "altexps_ERCC_percent", 
                     batch = sce.seger$Donor, 
                     subset = !sce.seger$Donor %in% c("HP1504901", "HP1509101"))

sce.seger <- sce.seger[, !(qc$discard | low.qual)]

# normalization
library(scran)
clusters <- quickCluster(sce.seger)
sce.seger <- computeSumFactors(sce.seger, clusters = clusters)
sce.seger <- logNormCounts(sce.seger)

library(scuttle)
pred.seger <- SingleR(test = sce.seger, 
                      ref = sce.murao, 
                      labels = sce.murao$label, 
                      de.method = "wilcox")

pred.seger$labels %>% table
```

As it so happens, we are in the fortunate position where our test dataset also contains independently defined labels. We see strong consistency between the two sets of labels (Figure 12.3), indicating that our automatic annotation is comparable to that generated manually by domain experts.  

```{r}
tab <- table(pred.seger$pruned.labels, sce.seger$CellType)

library(pheatmap)
library(viridis)
pheatmap(log2(tab + 1), color = viridis(100))
```

An interesting question is - given a single-cell reference dataset, is it better to use it directly or convert it to pseudo-bulk values? A single-cell reference preserves the “shape” of the subpopulation in high-dimensional expression space, potentially yielding more accurate predictions when the differences between labels are subtle (or at least capturing ambiguity more accurately to avoid grossly incorrect predictions). However, it also requires more computational work to assign each cell in the test dataset. We refer to [the other book](http://bioconductor.org/books/release/SingleRBook/) for more details on how to achieve a compromise between these two concerns.  

## Assigning cell labels from gene sets
A related strategy is to explicitly identify sets of marker genes that are highly expressed in each individual cell. This does not require matching of individual cells to the expression values of the reference dataset, which is faster and more convenient when only the identities of the markers are available. We demonstrate this approach using neuronal cell type markers derived from the Zeisel et al. (2015) study.  

```{r}
sce.zeisel <- readRDS("F:/git/Data/Bioconductor/osca/sce.zeisel.rds")

library(scater)
sce.zeisel <- aggregateAcrossFeatures(sce.zeisel, 
                                      id = sub("_loc[0-9]+$", "", rownames(sce.zeisel)))

# gene annotation
library(org.Mm.eg.db)
rowData(sce.zeisel)$ensembl <- mapIds(org.Mm.eg.db, 
                                      keys = rownames(sce.zeisel), 
                                      keytype = "SYMBOL", 
                                      column = "ENSEMBL")

# quality control
stats <- perCellQCMetrics(sce.zeisel, subsets = list(Mt = rowData(sce.zeisel)$featureType == "mito"))
qc <- quickPerCellQC(stats, percent_subsets = c("altexps_ERCC_percent", "subsets_Mt_percent"))
sce.zeisel <- sce.zeisel[, !qc$discard]

# normalization
library(scran)
set.seed(1000)
clusters <- quickCluster(sce.zeisel)
sce.zeisel <- computeSumFactors(sce.zeisel, clusters = clusters)
sce.zeisel <- logNormCounts(sce.zeisel)
```

```{r}
library(scran)

wilcox.z <- pairwiseWilcox(sce.zeisel, sce.zeisel$level1class, lfc = 1, direction = "up")
markers.z <- getTopMarkers(wilcox.z$statistics, wilcox.z$pairs, pairwise = F, n = 50)
markers.z %>% lapply(length) %>% unlist(use.names = T)
```

Our test data will be brain scRNA-seq experiment from Tasic et al.(2016).  

```{r}
library(scRNAseq)

sce.tasic <- TasicBrainData()
# saveRDS(sce.tasic, "F:/git/Data/Bioconductor/osca/sce.tasic.rds")
```

We use the AUCell package to identify marker sets that are highly expressed in each cell. This method ranks genes by their expression values within each cell and constructs a response curve of the number of genes from each marker set that are present with increasing rank. It then computes the area under the curve (AUC) for each marker set, quantifying the enrichment of those markers among the most highly expressed genes in that cell. This is roughly similar to performing a Wilcoxon rank sum test between genes in and outside of the set, but involving only the top ranking genes by expression in each cell.  






