---
title: "17 Cell cycle assignment"
author: "yincy"
date: "6/19/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Motivation
On occasion, it can be desirable to determine cell cycle activity from scRNA-seq data. In and of itself, the distribution of cells across phases of the cell cycle is not usually informative, but we can use this to determine if there are differences in proliferation between subpopulations or across treatment conditions. Many of the key events in the cell cycle (e.g., passage through checkpoints) are driven by post-translational mechanisms and thus not directly visible in transcriptomic data; nonetheless, there are enough changes in expression that can be exploited to determine cell cycle phase. We demonstrate using the 416B dataset, which is known to contain actively cycling cells after oncogene induction.  

```{r}
# loading
library(scRNAseq)
sce.416b <- LunSpikeInData(which = "416b")
sce.416b <- readRDS("/home/yincy/git/Data/Bioconductor/scRNAseq/sce.416b.rds")
sce.416b$block <- factor(sce.416b$block)

# gene annotation
library(AnnotationHub)
ens.mm.v97 <- AnnotationHub()[["AH73905"]]
rowData(sce.416b)$ENSEMBL <- rownames(sce.416b)
rowData(sce.416b)$SYMBOL <- mapIds(ens.mm.v97, 
                                   keys = rownames(sce.416b), 
                                   keytype = "GENEID", 
                                   column = "SYMBOL")
rowData(sce.416b)$SEQNAME <- mapIds(ens.mm.v97, 
                                    keys = rownames(sce.416b), 
                                    keytype = "GENEID", 
                                    column = "SEQNAME")

library(scater)
rownames(sce.416b) <- uniquifyFeatureNames(
    ID = rowData(sce.416b)$ENSEMBL, 
    names = rowData(sce.416b)$SYMBOL
)

# quality control
mito <- grepl("^MT$", rowData(sce.416b)$SEQNAME, ignore.case = F)
mito %>% table
rowData(sce.416b)$SEQNAME[mito]

stats <- perCellQCMetrics(x = sce.416b, subsets = list(Mito = mito))
qc <- quickPerCellQC(stats, 
                     percent_subsets = c("subsets_Mito_percent", "altexps_ERCC_percent"), 
                     batch = sce.416b$block)
sce.416b <- sce.416b[, !qc$discard]

# normalization
library(scran)
sce.416b <- computeSumFactors(sce.416b)
sce.416b <- logNormCounts(sce.416b)

# variance modelling
dec.416b <- modelGeneVarWithSpikes(sce.416b, spikes = "ERCC", block = sce.416b$block)
chosen.hvgs <- getTopHVGs(dec.416b, prop = 0.1)

# batch correction
library(limma)
assay(sce.416b, "corrected") <- removeBatchEffect(logcounts(sce.416b), 
                                                  design = model.matrix(~sce.416b$phenotype), 
                                                  batch = sce.416b$block)

# dimensionality reduction
sce.416b <- runPCA(sce.416b, 
                   ncomponents = 10, 
                   subset_row = chosen.hvgs, 
                   exprs_values = "corrected", 
                   BSPARAM = BiocSingular::ExactParam())

set.seed(1010)
sce.416b <- runTSNE(sce.416b, dimred = "PCA", perplexity = 10)

# clustering
my.dist <- dist(reducedDim(sce.416b, "PCA"))
my.tree <- hclust(my.dist, method = "ward.D2")

library(dynamicTreeCut)
my.clusters <- unname(cutreeDynamic(my.tree, distM = as.matrix(my.dist), 
                                    minClusterSize = 10, verbose = 0))
colLabels(sce.416b) <- factor(my.clusters)
sce.416b
```

## Using the cyclins
The cyclins control progression through the cell cycle and have well-characterized patterns of expression across cell cycle phases. Cyclin D is expressed throughout but peaks at G1; cyclin E is expressed highest in the G1/S transition; cyclin A is expressed across S and G2; and cyclin B is expressed highest in late G2 and mitosis (Morgan 2007). The expression of cyclins can help to determine the relative cell cycle activity in each cluster (Figure 17.1). For example, most cells in cluster 1 are likely to be in G1 while the other clusters are scattered across the later phases.  

```{r}
knitr::include_graphics("figures/fluction-of-cyclin-levels-during-cell-cycle.png")
```


```{r}
library(scater)

cyclin.genes <- grep("^Ccn[abde][0-9]{,1}$", rowData(sce.416b)$SYMBOL)
cyclin.genes <- rownames(sce.416b)[cyclin.genes]
```

```{r}
plotHeatmap(sce.416b, 
            order_columns_by = "label", 
            cluster_rows = F, 
            features = sort(cyclin.genes), 
            color = viridis::viridis(n = 100, option = "D"))
```

We quantify these observations with standard DE methods (Chapter 11) to test for upregulation of each cyclin between clusters, which would imply that a subpopulation contains more cells in the corresponding cell cycle phase. The same logic applies to comparisons between treatment conditions as described in Chapter 14. For example, we can infer that cluster 4 has the highest proportion of cells in the S and G2 phases based on higher expression of cyclins A2 and B1, respectively.  

```{r}
library(scran)

markers <- findMarkers(sce.416b, 
                       subset.row = cyclin.genes, 
                       test.type = "wilcox", 
                       direction = "up")

markers[[4]]
```

While straightforward to implement and interpret, this approach assumes that cyclin expression is unaffected by biological processes other than the cell cycle. This is a strong assumption in highly heterogeneous populations where cyclins may perform cell-type-specific roles. For example, using the Grun HSC dataset (Grun et al. 2016), we see an upregulation of cyclin D2 in sorted HSCs (Figure 17.2) that is consistent with a particular reliance on D-type cyclins in these cells (Steinman 2002; Kozar et al. 2004). Similar arguments apply to other genes with annotated functions in cell cycle, e.g., from relevant Gene Ontology terms.  

```{r}
# loading
library(scRNAseq)
sce.grun.hsc <- GrunHSCData(ensembl = TRUE)

# gene annotation
library(AnnotationHub)
ens.mm.v97 <- AnnotationHub()[["AH73905"]]
anno <- select(ens.mm.v97, 
               keys = rownames(sce.grun.hsc), 
               keytype = "GENEID", 
               columns = c("SYMBOL", "SEQNAME"))

rowData(sce.grun.hsc) <- anno[match(rownames(sce.grun.hsc), anno$GENEID), ]

# quality control
library(scater)
stats <- perCellQCMetrics(sce.grun.hsc)
qc <- quickPerCellQC(stats, batch = sce.grun.hsc$protocol, 
                     subset = grepl("sorted", sce.grun.hsc$protocol))
sce.grun.hsc <- sce.grun.hsc[, !qc$discard]

# normalization
library(scran)
set.seed(101000110)
clusters <- quickCluster(sce.grun.hsc)
sce.grun.hsc <- computeSumFactors(sce.grun.hsc, clusters = clusters)
sce.grun.hsc <- logNormCounts(sce.grun.hsc)

# variance modelling
set.seed(00010101)
dec.grun.hsc <- modelGeneVarByPoisson(sce.grun.hsc)
top.grun.hsc <- getTopHVGs(dec.grun.hsc, prop = 0.1)

# dimensional reduction
set.seed(101010011)
sce.grun.hsc <- denoisePCA(sce.grun.hsc, 
                           technical = dec.grun.hsc, 
                           subset.row = top.grun.hsc)
sce.grun.hsc <- runTSNE(sce.grun.hsc, dimred = "PCA")

# clustering
snn.gr <- buildSNNGraph(sce.grun.hsc, use.dimred = "PCA")
colLabels(sce.grun.hsc) <- factor(igraph::cluster_walktrap(snn.gr)$membership)
```

```{r}
rownames(sce.grun.hsc) <- uniquifyFeatureNames(rownames(sce.grun.hsc), 
                                               rowData(sce.grun.hsc)$SYMBOL)

cyclin.genes <- grep("^Ccn[abde]{1}[0-9]{1}$", rowData(sce.grun.hsc)$SYMBOL)
cyclin.genes <- rownames(sce.grun.hsc)[cyclin.genes]

plotHeatmap(sce.grun.hsc, 
            order_columns_by = "label", 
            colour_columns_by = "protocol", 
            cluster_row = F, 
            features = sort(cyclin.genes))
```

Admittedly, this is merely a symptom of a more fundamental issue - that the cell cycle is not independent of the other processes that are occurring in a cell. This will be a recurring theme throughout the chapter, which suggests that cell cycle inferences are best used in comparisons between closely related cell types where there are fewer changes elsewhere that might interfere with interpretation.  

## Using reference profiles
Cell cycle assignment can be considered a specialized case of cell annotation, which suggests that the strategies described in Chapter 12 can also be applied here. Given a reference dataset containing cells of known cell cycle phase, we could use methods like SingleR to determine the phase of each cell in a test dataset. We demonstrate on a reference of mouse ESCs from Buettner et al. (2015) that were sorted by cell cycle phase prior to scRNA-seq.  

```{r}
library(scRNAseq)
sce.ref <- BuettnerESCData()
sce.ref <- logNormCounts(sce.ref)
sce.ref
```

We will restrict the annotation process to a subset of genes with a priori known roles in cell cycle. This aims to avoid detecting markers for other biological processes that happen to be correlated with the cell cycle in the reference dataset, which would reduce classification performance if those processes are absent or uncorrelated in the test dataset.

```{r}
# find genes that are cell cycle-related
library(org.Mm.eg.db)
cycle.anno <- select(org.Mm.eg.db, 
                     keytype = "GOALL", 
                     keys = "GO:0007049", 
                     columns = "ENSEMBL")[, "ENSEMBL"]
cycle.anno %>% str
```

We use the `SingleR()` function to assign labels to the 416B data based on the cell cycle phases in the ESC reference. Cluster 1 mostly consists of G1 cells while the other clusters have more cells in the other phases, which is broadly consistent with our conclusions from the cyclin-based analysis. Unlike the cyclin-based analysis, this approach yields “absolute” assignments of cell cycle phase that do not need to be interpreted relative to other cells in the same dataset.  

```{r}
# Switch row names back to Ensembl to match the reference
test.data <- logcounts(sce.416b)
rownames(test.data) <- rowData(sce.416b)$ENSEMBL

library(SingleR)
assignments <- SingleR(test = test.data, 
                       ref = sce.ref, 
                       label = sce.ref$phase, 
                       de.method = "wilcox", 
                       restrict = cycle.anno)

tab <- table(assignments$labels, colLabels(sce.416b))
tab
```

The key assumption here is that the cell cycle effect is orthogonal to other aspects of biological heterogeneity like cell type. This justifies the use of a reference involving cell types that are quite different from the cells in the test dataset, provided that the cell cycle transcriptional program is conserved across datasets (Bertoli, Skotheim, and Bruin 2013; Conboy et al. 2007). However, it is not difficult to find holes in this reasoning - for example, Lef1 is detected as one of the top markers to distinguish between G1 from G2/M in the reference but has no detectable expression in the 416B dataset (Figure 17.3). More generally, non-orthogonality can introduce biases where, e.g., one cell type is consistently misclassified as being in a particular phase because it happens to be more similar to that phase’s profile in the reference.  

```{r}
library(patchwork)
plotExpression(sce.ref, features = "ENSMUSG00000027985", x = "phase") +
plotExpression(sce.416b, features = "Lef1", x = "label")
```

Thus, a healthy dose of skepticism is required when interpreting these assignments. Our hope is that any systematic assignment error is consistent across clusters and conditions such that they cancel out in comparisons of phase frequencies, which is the more interesting analysis anyway. Indeed, while the availability of absolute phase calls may be more appealing, it may not make much practical difference to the conclusions if the frequencies are ultimately interpreted in a relative sense (e.g., using a chi-squared test).  

```{r}
# test for differences in phase distributions between clusters 1 and 1.
chisq.test(tab[, 1:2])
```

## Using the `cyclone()` classifier
The method described by Scialdone et al. (2015) is yet another approach for classifying cells into cell cycle phases. Using a reference dataset, we first compute the sign of the difference in expression between each pair of genes. Pairs with changes in the sign across cell cycle phases are chosen as markers. Cells in a test dataset can then be classified into the appropriate phase, based on whether the observed sign for each marker pair is consistent with one phase or another. This approach is implemented in the `cyclone()` function from the `scran` package, which also contains pre-trained set of marker pairs for mouse and human data.

```{r}
set.seed(100)
library(scran)

mm.pairs <- readRDS(system.file("exdata", "mouse_cycle_markers.rds", package = "scran"))

# using ensembl IDs to match up with the annotation in 'mm.pairs'
assignments <- cyclone(x = sce.416b, 
                       pairs = mm.pairs, 
                       gene.names = rowData(sce.416b)$ENSEMBL)
```

The phase assignment result for each cell in the 416B dataset is shown below. For each cell, a higher score for a phase corresponds to a higher probability that the cell is in that phase. We focus on the G1 and G2/M scores as these are the most informative for classification.  

```{r}
plot(assignments$scores$G1, 
     assignments$scores$G2M, 
     xlab = "G1 score", 
     ylab = "G2/M score", 
     pch = 19)
```

Cells are classified as being in G1 phase if the G1 score is above 0.5 and greater than the G2/M score; in G2/M phase if the G2/M score is above 0.5 and greater than the G1 score; and in S phase if neither score is above 0.5. We see that the results are quite similar to those from `SingleR()`, which is reassuring.  

```{r}
table(assignments$phases, colLabels(sce.416b))
```

The same considerations and caveats described for the SingleR-based approach are also applicable here. From a practical perspective, `cyclone()` takes much longer but does not require an explicit reference as the marker pairs are already computed.  

## Removing cell cycle effects
### Comments
For some time, it was popular to regress out the cell cycle phase prior to downstream analyses like clustering. The aim was to remove uninteresting variation due to cell cycle, thus improving resolution of other biological processes. With the benefit of hindsight, we do not consider cell cycle adjustment to be necessary for routine applications. In most scenarios, the cell cycle is a minor factor of variation, secondary to stronger factors like cell type identity. Moreover, most strategies for removal run into problems when cell cycle activity varies across cell types or conditions; this is not uncommon with, e.g., increased proliferation of T cells upon activation (Richard et al. 2018), changes in cell cycle phase progression across development (Roccio et al. 2013) and correlations between cell cycle and fate decisions (Soufi and Dalton 2016). Nonetheless, we will discuss some approaches for mitigating the cell cycle effect in this section.  

### With linear regression and friends
Here, we treat each phase as a separate batch and apply any of the batch correction strategies described in Chapter 28.8. The most common approach is to use a linear model to simply regress out any effect associated with the assigned phases, as shown below in Figure 17.5 via `regressBatches()`. Similarly, any functions that support blocking can use the phase assignments as a blocking factor, e.g., `block=` in `modelGeneVarWithSpikes()`.  

```{r}
library(batchelor)
dec.nocycle <- modelGeneVarWithSpikes(sce.416b, 
                                      spikes = "ERCC", 
                                      block = assignments$phases)
reg.nocycle <- regressBatches(sce.416b, batch = assignments$phases)

set.seed(100011)
reg.nocycle <- runPCA(reg.nocycle, 
                      exprs_values = "corrected", 
                      subset_row = getTopHVGs(dec.nocycle, prop = 0.1))

# shape points by induction status
relabel <- c("onco", "WT")[factor(sce.416b$phenotype)]
scaled <- scale_shape_manual(values = c(onco = 4, WT = 19))

plotPCA(sce.416b, 
        colour_by = I(assignments$phases), 
        shape_by = I(relabel)) +
    ggtitle("Before") + 
    scaled +
plotPCA(reg.nocycle, colour_by = I(assignments$phases), 
        shape_by = I(relabel)) +
    ggtitle("After") +
    scaled
```

Alternatively, one could regress on the classification scores to account for any ambiguity in assignment. An example using `cyclone()` scores is shown below but the same procedure can be used with any classification step that yields some confidence per label - for example, the correlation-based scores from `SingleR()`.  

```{r}
design <- model.matrix(~ as.matrix(assignments$scores))
dec.nocycle2 <- modelGeneVarWithSpikes(sce.416b, spikes = "ERCC", design = design)
reg.nocycle2 <- regressBatches(sce.416b, design = design)

set.seed(100011)
reg.nocycle2 <- runPCA(reg.nocycle2, 
                       exprs_values = "corrected", 
                       subset_row = getTopHVGs(dec.nocycle2, prop = 0.1))

plotPCA(reg.nocycle2, 
        colour_by = I(assignments$phases), 
        point_size = 3, 
        shape_by = I(relabel)) +
    scaled
```

The main assumption of regression is that the cell cycle is consistent across different aspects of cellular heterogeneity (Section 13.4). In particular, we assume that each cell type contains the same distribution of cells across phases as well as a constant magnitude of the cell cycle effect on expression. Violations will lead to incomplete removal or, at worst, overcorrection that introduces spurious signal - even in the absence of any cell cycle effect! For example, if two subpopulations differ in their cell cycle phase distribution, regression will always apply a non-zero adjustment to all DE genes between those subpopulations.  

If this type of adjustment is truly necessary, it is safest to apply it separately to the subset of cells in each cluster. This weakens the consistency assumptions as we do not require the same behavior across all cell types in the population. Alternatively, we could use other methods that are more robust to differences in composition (Figure 17.7), though this becomes somewhat complicated if we want to correct for both cell cycle and batch at the same time. Gene-based analyses should use the uncorrected data with blocking where possible (Section 13.8), which provides a sanity check that protects against distortions introduced by the adjustment.  

```{r}
set.seed(100011)
reg.nocycle3 <- fastMNN(sce.416b, batch=assignments$phases)
plotReducedDim(reg.nocycle3, dimred="corrected", point_size=3,
    colour_by=I(assignments$phases), shape_by=I(relabel)) + scaled
```

### Removing cell cycle-related genes
















