---
title: "18 Trajectory Analysis"
author: "yincy"
date: "6/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Overview
Many biological processes manifest as a continuum of dynamic changes in the cellular state. The most obvious example is that of differentiation into increasingly specialized cell subtypes, but we might also consider phenomena like the cell cycle or immune cell activation that are accompanied by gradual changes in the cell's transcriptome. We characterize these processes from single-cell expression data by identifying a "trajectory", i.e., a path through the high-dimensional expression space that traverses the various cellular states associated with a continuous process like differentiation. In the simplest case, a trajectory will be a simple path from one point to another, but we can also observe more complex trajectories that branch to multiple endpoints.

The "pseudotime" is defined as the positioning of cells along the trajectory that quantifies the relative activity or progression of the underlying biological process. For example, the pseudotime for a differentiation trajectory might represent the degree of differentiation from a pluripotent cell to a terminal state where cells with larger pseudotime values are more differentiated. This metric allows us to tackle questions related to the global population structure in a more quantitative manner. **The most common application is to fit models to gene expression against the pseudotime to identify the genes responsible for generating the trajectory in the first place, especially around interesting branch events**.

In this section, we will demonstrate several different approaches to trajectory analysis using the haematopoietic stem cell (HSC) dataset from Nestorowa et al. (2016).

```{r}
# loading
library(scRNAseq)
sce.nest <- NestorowaHSCData()

# gene annotation
library(AnnotationHub)
library(EnsDb.Mmusculus.v79)
anno <- select(EnsDb.Mmusculus.v79, 
               keys = rownames(sce.nest), 
               keytype = "GENEID", 
               columns = c("SYMBOL", "SEQNAME"))

rowData(sce.nest) <- anno[match(rownames(sce.nest), anno$GENEID), ]

# quality control
library(scater)
stats <- perCellQCMetrics(sce.nest)
qc <- quickPerCellQC(stats, percent_subsets = "altexps_ERCC_percent")
sce.nest <- sce.nest[, !qc$discard]

# normalization
library(scran)
set.seed(101000110)
clusters <- quickCluster(sce.nest)
sce.nest <- computeSumFactors(sce.nest, clusters = clusters)
sce.nest <- logNormCounts(sce.nest)

# variance modelling
dec.nest <- modelGeneVarWithSpikes(sce.nest, spikes = "ERCC")
top.nest <- getTopHVGs(dec.nest, prop = 0.1)

# dimensionality reduction
set.seed(101010011)
sce.nest <- denoisePCA(sce.nest, technical = dec.nest, subset.row = top.nest)
sce.nest <- runTSNE(sce.nest, dimred = "PCA")

# clustering
snn.gr <- buildSNNGraph(sce.nest, use.dimred = "PCA")
colLabels(sce.nest) <- factor(igraph::cluster_walktrap(snn.gr)$membership)
```

## Obtaining pseudotime orderings
### Overview
The pseudotime is simply a number describing the relative position of a cell in the trajectory, where cells with larger values are consider to be "after" their counterparts with smaller values. Branched trajectories will typically be associated with multiple pseudotimes, one per path through the trajectory; these values are not usually comparable across paths. It is worth noting that "pseudotime" is a rather unfortunate term as it may not have much to do with real-life time. For example, one can imagine a continuum of stress states where cells move in either direction (or not) over time but the pseudotime simply describes the transition from one end of the continuum to the other. In trajectories describing time-dependent processes like differentiation, a cell's pseudotime value may be used as a proxy for its relative age, but only if directionality can be inferred (see Section 18.4).

The big question is how to identify the trajectory from high-dimensional expression data and map individual cells onto it. A massive variety of different algorithms are available for doing so (Saelens et al. 2019), and while we will demonstrate only a few specific methods below, many of the concepts apply generally to all trajectory inference strategies. A more philosophical question is whether a trajectory even exists in the dataset. **One can interpret a continuum of states as a series of closely related (but distinct) subpopulations, or two well-separated clusters as the endpoints of a trajectory with rare intermediates**. The choice between these two perspectives is left to the analyst based on which is more useful, convenient or biologically sensible.

### Cluster-based minimum spanning tree
#### Basic steps
**The `TSCAN` algorithm uses a simple yet effective approach to trajectory reconstruction. It uses the clustering to summarize the data into a smaller set of discrete units, computes cluster centroids by averaging the coordinates of its member cells, and then forms the minimum spanning tree (MST) across those centroids**. The MST is simply an undirected acyclic graph that passes through each centroid exactly once and is thus the most parsimonious structure that captures the transitions between clusters. We demonstrate below on the Nestorowa et al. (2016) dataset, computing the cluster centroids in the low-dimensional PC space to take advantage of data compaction and denoising (Chapter 9).

```{r}
library(scater)
by.cluster <- aggregateAcrossCells(sce.nest, ids = colLabels(sce.nest))
centroids <- reducedDim(by.cluster, "PCA")

# set cluster = NULL as we have already aggregated above
# BiocManager::install("TSCAN")
library(TSCAN)
mst <- createClusterMST(centroids, clusters = NULL)
```

For reference, we can draw the same lines between the centroids in a t-SNE plot (Figure 18.1). This allows us to identify interesting clusters such as those at bifurcations or endpoints. Note that the MST in `mst` was generated from distances in the PC space and is merely being visualized here in the t-SNE space, for the same reasons as discussed in Section 9.5.5. This may occasionally result in some visually unappealing plots if the original ordering of clusters in the PC space is not preserved in the t-SNE space.  

```{r}
line.data <- reportEdges(by.cluster, mst = mst, clusters = NULL, use.dimred = "TSNE")

plotTSNE(sce.nest, 
         colour_by = "label") +
  geom_line(data = line.data, aes(dim1, dim2, group = edge))
```

We obtain a pseudotime ordering by projecting the cells onto the MST with `mapCellsToEdges()`. More specifically, we move each cell onto the closest edge of the MST; the pseudotime is then calculated as the distance along the MST to this new position from a "root node" with `orderCells()`. For our purposes, we will arbitrarily pick one of the endpoint nodes as the root, though a more careful choice based on the biological annotation of each node may yield more relevant orderings (e.g., picking a node corresponding to a more pluripotent state).  

```{r}
map.tscan <- mapCellsToEdges(x = sce.nest, mst = mst, use.dimred = "PCA")
tscan.pseudo <- orderCells(mapping = map.tscan, mst = mst)
tscan.pseudo %>% head
```

Here, multiple sets of pseudotimes are reported for a branched trajectory. Each column contains one pseudotime ordering and corresponds to one path from the root node to one of the terminal nodes - the name of the terminal node that defines this path is recorded in the column names of `tscan.pseudo`. Some cells may be shared across multiple paths, in which case they will have the same pseudotime in those paths. We can then examine the pseudotime ordering on our desired visualization.  

```{r}
# taking the rowMeans just gives us a single-pseudo-time for all cells. Cells in segments that are shared across paths have the same pseudo-time value for those paths anyway, so the rowMeans doesn't change angthing.  

common.pseudo <- rowMeans(tscan.pseudo, na.rm = T)
plotTSNE(sce.nest, 
         colour_by = I(common.pseudo), 
         text_by = "label", 
         text_colour = "red") +
  geom_line(data = line.data, aes(dim1, dim2, group = edge))
```

Alternatively, this entire series of calculations can be conveniently performed with the `quickPseudotime()` wrapper. This executes all steps from `aggregateAcrossCells()` to `orderCells()` and returns a list with the output from each step.  

```{r}
pseudo.all <- quickPseudotime(sce.nest, use.dimred = "PCA")
pseudo.all$ordering %>% head
```


#### Tweaking the MST
The MST can be constructed with an “outgroup” to avoid connecting unrelated populations in the dataset. Based on the OMEGA cluster concept from Street et al. (2018), the outgroup is an artificial cluster that is equidistant from all real clusters at some threshold value. If the original MST sans the outgroup contains an edge that is longer than twice the threshold, the addition of the outgroup will cause the MST to instead be routed through the outgroup. We can subsequently break up the MST into subcomponents (i.e., a minimum spanning forest) by removing the outgroup. We set `outgroup=TRUE` to introduce an outgroup with an automatically determined threshold distance, which breaks up our previous MST into two components.  

```{r}
pseudo.og <- quickPseudotime(sce.nest, use.dimred = "PCA", outgroup = TRUE)
set.seed(10101)
plot(pseudo.og$mst)
```

Another option is to construct the MST based on distances between mutual nearest neighbor (MNN) pairs between clusters (Section 13.5). This exploits the fact that MNN pairs occur at the boundaries of two clusters, with short distances between paired cells meaning that the clusters are “touching”. In this mode, the MST focuses on the connectivity between clusters, which can be different from the shortest distance between centroids. Consider, for example, a pair of elongated clusters that are immediately adjacent to each other. A large distance between their centroids precludes the formation of the obvious edge with the default MST construction; in contrast, the MNN distance is very low and encourages the MST to create a connection between the two clusters.  

```{r}
pseudo.mnn <- quickPseudotime(sce.nest, use.dimred = "PCA", with.mnn = TRUE)
mnn.pseudo <- rowMeans(pseudo.mnn$ordering, na.rm = T)
plotTSNE(sce.nest, 
         colour_by = I(mnn.pseudo), 
         text_by = "label", 
         text_colour = "red") +
  geom_line(data = pseudo.mnn$connected$TSNE, aes(dim1, dim2, group = edge))
```

#### Further comments
The `TSCAN` approach derives several advantages from using clusters to form the MST. The most obvious is that of computational speed as calculations are performed over clusters rather than cells. The relative coarseness of clusters protects against the per-cell noise that would otherwise reduce the stability of the MST. The interpretation of the MST is also straightforward as it uses the same clusters as the rest of the analysis, allowing us to recycle previous knowledge about the biological annotations assigned to each cluster.  

However, the reliance on clustering is a double-edged sword. If the clusters are not sufficiently granular, it is possible for `TSCAN` to overlook variation that occurs inside a single cluster. The MST is obliged to pass through each cluster exactly once, which can lead to excessively circuitous paths in overclustered datasets as well as the formation of irrelevant paths between distinct cell subpopulations if the outgroup threshold is too high. The MST also fails to handle more complex events such as “bubbles” (i.e., a bifurcation and then a merging) or cycles.  

### Principal curves
To identify a trajectory, one might imagine simply “fitting” a one-dimensional curve so that it passes through the cloud of cells in the high-dimensional expression space. This is the idea behind principal curves (Hastie and Stuetzle 1989), effectively a non-linear generalization of PCA where the axes of most variation are allowed to bend. We use the `slingshot` package (Street et al. 2018) to fit a single principal curve to the Nestorowa dataset, again using the low-dimensional PC coordinates for denoising and speed. This yields a pseudotime ordering of cells based on their relative positions when projected onto the curve.  

```{r}
library(slingshot)
sce.sling <- slingshot(sce.nest, reducedDim = "PCA")
colData(sce.sling)$slingPseudotime_1 %>% head
```

We can then visualize the path taken by the fitted curve in any desired space with embedCurves(). For example, Figure 18.5 shows the behavior of the principle curve on the t-SNE plot. Again, users should note that this may not always yield aesthetically pleasing plots if the t-SNE algorithm decides to arrange clusters so that they no longer match the ordering of the pseudotimes.

```{r}
embedded <- embedCurves(sce.sling, newDimRed = "TSNE")
embedded <- slingCurves(embedded)[[1]] # only one path
embedded <- data.frame(embedded$s[embedded$ord, ])

plotTSNE(sce.sling, 
         colour_by = "slingPseudotime_1") +
  geom_path(data = embedded, aes(Dim.1, Dim.2), size = 1)
```

The previous call to `slingshot()` assumed that all cells in the dataset were part of a single curve. To accommodate more complex events like bifurcations, we use our previously computed cluster assignments to build a rough sketch for the global structure in the form of a MST across the cluster centroids. Each path through the MST from a designated root node is treated as a lineage that contains cells from the associated clusters. Principal curves are then simultaneously fitted to all lineages with some averaging across curves to encourage consistency in shared clusters across lineages. This process yields a matrix of pseudotimes where each column corresponds to a lineage and contains the pseudotimes of all cells assigned to that lineage.  

```{r}
sce.sling2 <- slingshot(sce.nest, clusterLabels = colLabels(sce.nest), reducedDim = "PCA")
pseudo.paths <- slingPseudotime(sce.sling2)

pseudo.paths %>% head
```

By using the MST as a scaffold for the global structure, `slingshot()` can accommodate branching events based on divergence in the principal curves (Figure 18.6). However, unlike TSCAN, the MST here is only used as a rough guide and does not define the final pseudotime.  

```{r}
sce.nest <- runUMAP(sce.nest, dimred = "PCA")
reducedDim(sce.sling2, "UMAP") <- reducedDim(sce.nest, type = "UMAP")
shared.pseudo <- rowMeans(pseudo.paths, na.rm = T)

# need to loop over the paths and add each one separately
gg <- plotUMAP(sce.sling2, colour_by = I(shared.pseudo))
embedded <- embedCurves(sce.sling2, newDimRed = "UMAP")
embedded <- slingCurves(embedded)

for(path in embedded){
  embedded <- data.frame(path$s[path$ord, ])
  gg <- gg + geom_path(data = embedded, aes(Dim.1, Dim.2), size = 1)
}

gg
```

We can use `slingshotBranchID()` to determine whether a particular cell is shared across multiple curves or is unique to a subset of curves (i.e., is located “after” branching). In this case, we can see that most cells jump directly from a global common segment (`1,2,3`) to one of the curves (`1`, `2`, `3`) without any further hierarchy, i.e., no noticeable internal branch points.  

```{r}
curve.assignments <- slingBranchID(sce.sling2)

curve.assignments %>% table
```

For larger datasets, we can speed up the algorithm by approximating each principal curve with a fixed number of points. By default, `slingshot()` uses one point per cell to define the curve, which is unnecessarily precise when the number of cells is large. Applying an approximation with `approx_points=` reduces computational work without any major loss of precision in the pseudotime estimates.  

```{r}
sce.sling3 <- slingshot(sce.nest, 
                        clusterLabels = colLabels(sce.nest), 
                        reducedDim = "PCA", 
                        approx_points = 100)

pseudo.path3 <- slingPseudotime(sce.sling3)
pseudo.path3 %>% head
```

The MST can also be constructed with an OMEGA cluster to avoid connecting unrelated trajectories. This operates in the same manner as (and was the inspiration for) the outgroup for `TSCAN`’s MST. Principal curves are fitted through each component individually, manifesting in the pseudotime matrix as paths that do not share any cells.  

```{r}
sce.sling4 <- slingshot(sce.nest, 
                        cluster = colLabels(sce.nest), 
                        reducedDim = "PCA", 
                        approx_points = 100, 
                        omega = TRUE)

pseudo.path4 <- slingPseudotime(sce.sling4)
pseudo.path4 %>% head
```

```{r}
shared.pseudo <- rowMeans(pseudo.path4, na.rm = TRUE)
gg <- plotUMAP(sce.sling4, colour_by = I(shared.pseudo))
embedded <- embedCurves(sce.sling4, "UMAP")
embedded <- slingCurves(embedded)

for(path in embedded){
  embedded <- data.frame(path$s[path$ord, ])
  gg <- gg + geom_path(data = embedded, aes(Dim.1, Dim.2), size = 1)
}

gg
```

The use of principal curves adds an extra layer of sophistication that complements the deficiencies of the cluster-based MST. The principal curve has the opportunity to model variation within clusters that would otherwise be overlooked; for example, `slingshot` could build a trajectory out of one cluster while `TSCAN` cannot. Conversely, the principal curves can “smooth out” circuitous paths in the MST for overclustered data, ignoring small differences between fine clusters that are unlikely to be relevant to the overall trajectory.  

That said, the structure of the initial MST is still fundamentally dependent on the resolution of the clusters. One can arbitrarily change the number of branches from `slingshot` by tuning the cluster granularity, making it difficult to use the output as evidence for the presence/absence of subtle branch events. If the variation within clusters is uninteresting, the greater sensitivity of the curve fitting to such variation may yield irrelevant trajectories where the differences between clusters are masked. Moreover, `slingshot` is no longer obliged to separate clusters in pseudotime, which may complicate intepretation of the trajectory with respect to existing cluster annotations.  


## Characterizing trajectories
### Overview
Once we have constructed a trajectory, the next step is to characterize the underlying biology based on its DE genes. The aim here is to find the genes that exhibit significant changes in expression across pseudotime, as these are the most likely to have driven the formation of the trajectory in the first place. The overall strategy is to fit a model to the per-gene expression with respect to pseudotime, allowing us to obtain inferences about the significance of any association. We can then prioritize interesting genes as those with low p-values for further investigation. A wide range of options are available for model fitting but we will focus on the simplest approach of fitting a linear model to the log-expression values with respect to the pseudotime; we will discuss some of the more advanced models later.

### Changes along a trajectory
To demonstrate, we will identify genes with significant changes with respect to one of the `TSCAN` pseudotimes in the Nestorowa data. We use the `testPseudotime()` utility to fit a natural spline to the expression of each gene, allowing us to model a range of non-linear relationships in the data. We then perform an analysis of variance (ANOVA) to determine if any of the spline coefficients are significantly non-zero, i.e., there is some significant trend with respect to pseudotime.  

```{r}
library(TSCAN)
pseudo <- testPseudotime(sce.nest, pseudotime = tscan.pseudo[, 1])
pseudo$SYMBOL <- rowData(sce.nest)$SYMBOL
pseudo[order(pseudo$p.value), ]
```

In practice, it is helpful to pair the spline-based ANOVA results with a fit from a much simpler model where we assume that there exists a linear relationship between expression and the pseudotime. This yields an interpretable summary of the overall direction of change in the `logFC` field above, complementing the more poweful spline-based model used to populate the `p.value` field. In contrast, the magnitude and sign of the spline coefficients cannot be easily interpreted.  

To simplify the results, we will repeat our DE analysis after filtering out cluster 7. This cluster seems to contain a set of B cell precursors that are located at one end of the trajectory, causing immunoglobulins to dominate the set of DE genes and mask other interesting effects. Incidentally, this is the same cluster that was split into a separate component in the outgroup-based MST.  

```{r}
# making a copy of the sce.nest and including the pseudotimes in the colData
sce.nest2 <- sce.nest
sce.nest2$TSCAN.first <- tscan.pseudo[, 1]
sce.nest2$TSCAN.second <- tscan.pseudo[, 2]

# discarding the offending cluster
discard <- 7
keep <- colLabels(sce.nest) != 7
sce.nest2 <- sce.nest2[, keep]

# testing against the first path again
pseudo <- testPseudotime(sce.nest2, pseudotime = sce.nest2$TSCAN.first)
pseudo$symbol <- rowData(sce.nest2)$SYMBOL
sorted <- pseudo[order(pseudo$p.value), ]
```

Examination of the top downregulated genes suggests that this pseudotime represents a transition away from myeloid identity, based on the decrease in expression of genes such as Mpo and Plac8.  

```{r}
up.left <- sorted[sorted$logFC < 0, ]
up.left %>% head(10)
```

```{r}
best <- up.left$symbol %>% head(10)
plotExpression(sce.nest2, 
               features = best, 
               swap_rownames = "SYMBOL", 
               x = "TSCAN.first", 
               colour_by = "label", 
               ncol = 4) +
  theme(legend.position = "none")
```

Conversely, the later parts of the pseudotime may correspond to a more stem-like state based on upregulation of genes like Hlf. There is also increased expression of genes associated with the lymphoid lineage (e.g., Ltb), consistent with reduced commitment to the myeloid lineage at earlier pseudotime values.  

```{r}
up.right <- sorted[sorted$logFC > 0, ]
up.right %>% head(10)
```

```{r}
best <- up.right$symbol %>% head(10)
plotExpression(sce.nest2, features = best, 
               swap_rownames = "SYMBOL", 
               x = "TSCAN.first", 
               colour_by = "label", 
               ncol = 4) +
  theme(legend.position = "none")
```

```{r}
on.first.path <- !is.na(sce.nest2$TSCAN.first)
plotHeatmap(sce.nest2[, on.first.path], 
            order_columns_by = "TSCAN.first", 
            colour_columns_by = "label", 
            features = head(up.right$symbol, 50), 
            center = TRUE, 
            swap_rownames = "SYMBOL")
```

### Changes between paths














































