---
title: "An Introduction to ShortRead"
author: "yincy"
date: "11/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load package, message=FALSE}
library(ShortRead)
library(magrittr)
```


## Overview  
The `ShortRead` package contains functionality for reading and examining raw sequence reads (typically in FASTQ format).  

## ShortRead  
The `ShortRead` package was one of the first Bioconductor packages to deal with low-level analysis of high-throughput sequencing data. Some of its functionality has now been superseded by other packages, but there is still relevant functionality left.  

Functionality  

**Input**  
- `FastqStreamer`: Iterate through FASTQ files in chunks  
- `FastqSampler`: Draw random samples from FASTQ files  
- `readFastq`: Read an entire FASTQ file into memory  
- `writeFastq`: Write FASTQ objects to a connection (file)  
- `countFastq`: Quickly count FASTQ records in files  

**Sequence and quality summary**  
- `alphabetFrequency`: Nucleotide or quality score use per read  
- `alphabetByCycle`: Nucleotide or quality score use by cycle  
- `alphabetScore`: Whole-read quality summary  
- `encoding`: Character / 'phred' score mapping  

**Filtering and trimming**  
- `srFilter`: Pre-defined and bespoke filters  
- `trimTails, etc.`: Trim low-quality nucleotides  
- `narrow`: Remove leading / trailing nucleotides  
- `tables`: Summarize read occurrence  
- `srduplicated, etc.`: Identify duplicate reads  
- `filterFastq`: Filter reads from one file to another  


### Reading FASTQ files  
The FASTQ file format is the standard way of representing raw (unaligned) next generation sequencing reads, particular for the Illumina platform. The format basically consists of 4 lines per read, with the lines containing  
- Read name (sometimes includes flowcell ID or other information).  
- Read nucleotides  
- Either empty or a repeat of line 1  
- Encoded read quality scores  

Paired-end reads are stored in two separate files, where the reads are ordered the same (this is obviously fragile; what if reads are re-ordered in one file and not the other).  

These files are read by `readFastq()` which produces an object of class `ShortReadQ`  
```{r}
fastqDir <- system.file("extdata", "E-MTAB-1147", package = "ShortRead")
fastqPath <- list.files(fastqDir, pattern = ".fastq.gz$", full = TRUE)[1]
reads <- readFastq(dirPath = fastqPath)
```

Here we directly point the function to the file path. A paradigm which is often used in Bioconductor is to first put the file path into an object which represents a specific file type and then read it  

```{r}
fqFile <- FastqFile(fastqPath)
```

```{r}
reads <- readFastq(fqFile)
```


This appears to make little sense in this situation, but for really big files it makes sense to access them in chunks  

The `ShortReadQ` class is very similar to a `DNAStringSet` but it has two sets of strings: one for the read nucleotides and one for the base qualities.  

The `ShortReadQ` class contains:  

- `DNAStringSet`: (Biostrings) Short read sequences  
- `FastqQuality, etc`: Quality encodings  
- `ShortReadQ`: Reads, quality scores, and ids  


```{r}
sread(reads)
```

```{r}
quality(reads)
```

```{r}
id(reads)
```

```{r}
quality(reads) %>% encoding()
```


### A word on quality scores  
convert quality scores to integer  
```{r}
quality(reads) %>% as(Class = "matrix") %>% .[1:5, 1:5]
```

In this conversion, each letter is matched to an integer between 0 and 40. This matching is known as the “encoding” of the quality scores and there has been different ways to do this encoding. Unfortunately, it is not stored in the FASTQ file which encoding is used, so you have to know or guess the encoding. The `ShortRead` package does this for you.  


## Common work flows
### Quality assessment
FASTQ files are often used for basic quality assessment, often to augment the purely technical QA that might be provided by the sequencing center with QA relevant to overall experimental design. A QA report is generated by creating a vector of paths to FASTQ files.  

```{r}
fls <- dir("../../data/kidney/kidney-map/glomer2021/fastq/", full.names = T)
fls %>% head

qaSummary <- qa(fls[1:100], type = "fastq")
```

creating and viewing a report  

```{r}
browseURL(report(qaSummary))
```


## Filtering and trimming
```{r}
myFilterAndTrim
```

























