---
title: "GenomicRanges"
author: "yincy"
date: "11/6/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load packages, message=FALSE}
library(GenomicRanges)
library(magrittr)
```

## GRanges  
`GRanges` are like `IRanges` with strand and chromosome. Strand can be +, - and *. The value * indicates 'unknown strand' or 'unstranded'.  

This value usually gets treated as a third strand, which is sometimes confusing to users.  

```{r}
?GRanges
gr <- GRanges(seqnames = "chr1", 
              strand = c("+", "-", "+"),
              ranges = IRanges(start = c(1, 3, 5), width = 3))
```

Natural accessor functions: `strand()`, `seqnames()`, `ranges()`, `start()`, `end()`, `width()`.  

Because the they have strand, we now have operations which are relative to the direction of transcription: `upstream()`, `downstream()`.  

```{r}
gr
flank(gr, width = 2, start = T); flank(gr, width = 2, start = F)
```


```{r}
strand(gr)
start(gr)
end(gr)
width(gr)
seqnames(gr)
```


## GRanges, seqinfo  
`GRanges` operate within a universe of sequences (chromosome/contigs) and their lengths.  
```{r}
seqinfo(gr)
```

```{r}
seqlengths(gr) <- c("chr1" = 10)
seqinfo(gr)
```

```{r}
seqlevels(gr)
```

```{r}
seqlengths(gr)
```

Especially the length of the chromosomes are used by some functions. For example `gaps()` return the stretches of the genome not covered by the `GRanges`.  
```{r}
gaps(x = gr, start = 1L, end = seqlengths(gr));gr
length(gr)
```

In this example, we known that the last gap stops at 10, because that is the length of the chromosome.  

Note how a range on the * strand appears in the result.  

```{r}
seqlevels(gr) <- c("chr1", "chr2")
seqnames(gr) <- c("chr1", "chr2", "chr1")
```

When you `sort()` a `GRanges`, the sorting order of the chromosomes is determined by their order in `seqlevel`.   
```{r}
sort(gr)
```

```{r}
seqlevels(gr) <- c("chr2", "chr1")
sort(gr)
```

associate a genome with a `GRanges`  
```{r}
genome(gr) <- "hg19"
gr
```

This is becomes valuable when you deal with data from different genome versions, because it allows R to throw an error when you compare two `GRanges` from different genomes   
```{r}
gr2 <- gr
genome(gr2) <- "hg18"
findOverlaps(gr, gr2)
```

## DataFrame  
The `S4Vectors` package introduced the **DataFrame** class. This class is very similar to the base `data.frame` class from R, but it allows columns of any class, provided a number of required methods are supported. For example, **DataFrame** can have `IRanges` as columns, unlike `data.frame`.  

```{r}
ir <- IRanges(start = 1:2, width = 3)
df1 <- DataFrame(iranges = ir)
```

```{r}
df1$iranges
```

```{r}
df2 <- data.frame(iranges = ir)
```

## Granges, metadata  
`GRanges` (unlike `IRanges`) may have associated metadata. This is immensely useful. The formal way to access and set this metadata is through `values` or `elementMetadata` or `mcols`  

```{r}
gr <- GRanges(seqnames = "chr1", 
              strand = c("+", "-", "+"),
              ranges = IRanges(start = c(1, 3, 5), width = 3))

values(gr) <- DataFrame(score = c(0.1, 0.5, 0.3))
gr
```

A much easier way to set and access metadata is through the $ operator  
```{r}
gr$score
```


```{r}
gr$score2 <- gr$score * 2
gr
```

## findOverlaps  
`findOverlas` works exactly as for `IRanges`. But the `strand` information can be confusing.  
```{r}
gr2 <- GRanges(seqnames = c("chr1", "chr2", "chr1"),
               strand = "*",
               ranges = IRanges(start = c(1, 3, 5), width = 3))
gr2
```

```{r}
gr
```

```{r}
findOverlaps(gr, gr2)
```

Notice how the * strand overlaps both + and -. There is an argument *ignore.strand* to `findOverlaps` which will ignore the strand information (so + overlaps -). Several other functions in `GenomicRanges` have an *ignore.strans* argument as well.  

```{r}
findOverlaps(gr, gr2, ignore.strand = T)
```


## subsetByOverlaps  
A common operation is to select only certain ranges from a `GRanges` which overlap something else.   
```{r}
gr;gr2
subsetByOverlaps(gr, gr2)
```


## makeGRangesFromeDataFrame  
A common situation is that you have data which looks like a `GRanges` but is really stored as a classic `data.frame`, with chr, start etc. The `makeGRangesFromDataFrame` converts this `data.frame` into a `GRanges`. An argument tells you whether you want to keep any additional columns.  
```{r}
df <- data.frame(chr = "chr1", 
                 start = 1:3, 
                 end = 4:6,
                 score = 7:9)
makeGRangesFromDataFrame(df)
```

```{r}
makeGRangesFromDataFrame(df, keep.extra.columns = T)
```

## seqinfo 
The `GRanges` class contains `seqinfo` information about the length and the names of the chromosomes. Here we will briefly discuss strategies for harmonizing this information.  

The `GenomeInfoDb` package addresses a seemingly small, but consistent problem: different online resources uses different naming conventions for chromosomes. In more technical jargon, this package helps keeping your `seqinfo` and `seqlevels` harmonized.  

## Drop and keep seqlevels  
It is common to want to remove `seqlevels` from a `GRanges` object.   
```{r}
gr <- GRanges(seqnames = c("chr1", "chr2"),
              ranges = IRanges(start = 1:2, end = 4:5))

seqlevels(gr) <- c("chr1", "chr2")
```

In `GenomeInfoDb` you find `dropSeqlevels()` and `keepSeqlevels`.  
```{r}
gr <- GRanges(seqnames = c("chr1", "chr2"),
              ranges = IRanges(start = 1:2, end = 4:5))
```

```{r}
dropSeqlevels(x = gr, value = "chr1", pruning.mode = "coarse")
```


```{r}
keepSeqlevels(x = gr, value = "chr2", pruning.mode = "coarse")
```

get rid of weired looking chromosome names with `keepStandardChromosomes()`  
```{r}
gr <- GRanges(seqnames = c("chr1", "chrU345"),
              ranges = IRanges(start = 1:2, end = 4:5))

keepStandardChromosomes(gr, pruning.mode = "coarse")
```


## Changing style  
It is an inconvenient truth that different online resources uses different naming convention for chromosomes. This can even be different from organism to organism. For example, for the fruitfly (Drosophila Melanogaster) NCBI and Ensembl uses “2L” and UCSC uses “chr2L”. But NCBI and Ensembl differs on some contigs: NCBI uses “Un” and Ensembl used “U”.  

```{r}
gr <- GRanges(seqnames = "chr1", ranges = IRanges(start = 1:2, width = 2))
```


```{r remap}
newStyle <- mapSeqlevels(seqnames = seqlevels(gr),style = "NCBI" )

gr <- renameSeqlevels(gr, value = newStyle)
```

This can in principle go wrong, if the original set of `seqlevels` are inconsistent (not a single style).  


# GenomicRanges - Rle
`Rle`: run length encoding, this class is great for representation genome-wide sequence coverage.  

## Coverage 
In high-throughput sequencing, coverage is the number of reads overlapping each base. In other words, it associates a number (the number of reads) to each base in the genome.  

This is fundamental quantity for many high-throughput sequencing analyses. For variant calling (DNA-sequencing) it tells you how much power (information) you have to call a variant at a given location.  

For ChIP sequencing it is the primary signal; areas with high coverage are thought to be enriched for a given protein.  

## Rle  
An `Rle` (run-length-encoded) vector is a specific representation of a vector. The `IRanges` package implements support for this class.  

Note: there is also a base class called `rle` which has much less functionality.  

The run-length-encoded representation of a vector, represents the vector as a set of distinct runs with their own value.  

```{r}
rl <- Rle(values = c(1, 1, 1, 1, 2, 2, 3, 3, 2, 2))
```

```{r}
runLength(rl)
```


```{r}
runValue(rl)
```

```{r}
as.numeric(rl)
```

This is very efficient representation if:  
- the vector is very long  
- there are a lot of consecutive elemnts with the same value  

This is especially useful for genomic data which is either piece wise constant, or where most of the genome is not covered.  

In many ways `Rle`s function as normal vectors, you can do arithmetic with them, transform them etc. using standard R functions like `+` and `log2`.  

There are also `RleList` which is a list of `Rle`s. This class is used to represent a genome wide coverage track where each element of the list is a different chromosome.  


## Useful functions for Rle  
A standard usecase is that you have a number of regions (say `IRanges`) and you want to do something to your `Rle` over each of these regions.   
```{r}
ir <- IRanges(start = c(2, 6), width = 2)
aggregate(x = rl, by = ir, FUN = max)
```

```{r}
ir <- IRanges(start = 1:10, width = 3)
rl <- coverage(ir)
```

select high coverage regions by the `slice()` function  
```{r}
slice(rl, lower = 2)
```



## Views and Rles  
```{r}
rl
vi <- Views(subject = rl, start = c(3, 7), width = 3)
vi
```

```{r}
mean(vi)
```


## RleList  
An `RleList` is simply a list of `Rle`. It is similar to a `GRangesList` in concept.  


## Rles and GRanges  
`Rle`'s can also be constructed fro `GRanges`.  

```{r}
gr <- GRanges(seqnames = "chr1", 
              ranges = IRanges(start = 1:10, width = 3))

rl <- coverage(gr)
rl
```


# GenomicRanges - Lists  
An important usecase specifically for `GRangesList` is the representation of a set of **transcripts**. Each transcript is an element in the list and the exons of the transcript is represented as a `GRanges`.  

```{r}
gr1 <- GRanges(seqnames = "chr1", ranges = IRanges(start = 1:4, width = 3))
gr2 <- GRanges(seqnames = "chr2", ranges = IRanges(start = 1:4, width = 3))

gL <- GRangesList(gr1 = gr1, gr2 = gr2)
```

A number of standard `GRanges` functions work, but returns (for example) `IntegerList`s  
```{r}
start(gL)
```

```{r}
seqnames(gL)
```


```{r}
sapply(gL, length)
```

We have a new `XXapply` function with the fancy name `endoapply`. This is used when you want to apply a function which maps a `GRanges` into a `GRanges`  
```{r}
shift(gL, 10)
```

`findOverlaps` works slightly different. For `GRangesLists`, we think of each element is a union of ranges. So we get an overlap if any range overlaps.  
```{r}
findOverlaps(gL, gr2)
```

Note how the `queryLength` is 2 and not 20. What we know from the first row of this output is that some range in gL[[2]] overlaps the range gr[1].

This is actually a feature if we think of the `GRangesList` as a set of transcript, where each `GRanges` gives you the exon of the transcript. With this interpretation, `findOverlaps` tells you whether or not the transcript overlaps some region of interest, and this is true if any of the exons of the transcript overlaps the region.  


## Other Lists  
There are many other types of `XXList`  
- `RleList`  
- `IRangesList`  
- `IntegerList`  
- `CharacterList`  
- `LogicalList`  

and many others.  


# From Vignette 
## An introduction to the GenomicRanges package  
### Introduction  
The `GenomicRanges` package serves as the foundation for representing genomic locations within the Bioconductor project. In the Bioconductor package hierarchy, it builds upon the `IRanges` (infrastructure) package and provides support for the `BSgenome` (infrastructure), `Rsamtools` (I/O), `ShortRead` (I/O & QA), `rtracklayer` (I/O), `GenomicFeatures` (infrastructure), `GenomicAlignments` (sequence reads), `VariantAnnotation` (called variants), and many other Bioconductor packages.  

This package lays a foundation for genomic analysis by introducing three classes (*GRanges*, *GPos*, and *GRangesList*), which are used to represent genomic ranges, genomic positions, and groups of genomic ranges.  


### GRanges: Genomic Ranges  
The GRanges class represents a collection of genomic ranges that each have a single start and end location on the genome. It can be used to store the location of genomic features such as contiguous binding sites, transcripts, and exons.  

```{r}
gr <- GRanges(
    seqnames = Rle(c("chr1", "chr2", "chr1", "chr3"), c(1, 3, 2, 4)),
    ranges = IRanges(101:110, end = 111:120, names = head(letters, 10)),
    strand = Rle(strand(c("-", "+", "*", "+", "-")), c(1, 2, 2, 3, 2)),
    score = 1:10,
    GC = seq(1, 0, length=10))
```

The output of the `GRanges` show method separates the information into a left and right hand region that are separated by `|` symbols. The genomic coordinates (seqnames, ranges, and strand) are located on the left-hand side and the metadata columns (annotation) are located on the right.  

```{r}
gr
seqnames(gr)
ranges(gr) # out IRanges object
range(gr) # output GRanges object  
granges(gr) # output GRanges object 
strand(gr)
mcols(gr)
mcols(gr)$score
seqlengths(gr) <- c(249250621, 243199373, 198022430)
names(gr)
```


```{r}
sp <- split(gr, f = rep(1:2, each = 5))

c(sp[[1]], sp[[2]])
```


# GenomicRanges HOWTOs  
## How to read single-end reads from a BAM file  
Several functions are available for reading BAM files into R  
- `readGAlignments()`  
- `readGAlignmentPairs()`  
- `readGAlignmentsList()`  
- `scanBam()`  

`scanBam()` in `Rsamtools` is a low-level function that returns a list of lists.  

Single-end reads can be loaded with the `readGAlignments` function from the `GenomicAlignments` package.  

```{r}
library(pasillaBamSubset)
un1 <- untreated1_chr4() # single-end reads 
```


```{r}
library(GenomicAlignments)
gal <- readGAlignments(un1)
```


We input records that overlap position 1 to 5000 on the negative strand with `flag` and `cigar` as metadata columns.  
```{r}
what <- c("flag", "cigar")
which <- GRanges(seqnames = "chr4", 
                 ranges = IRanges(start = 1, end = 5000))

flag <- scanBamFlag(isMinusStrand = TRUE)
param <- ScanBamParam(flag = flag, what = what, which = which)
neg <- readGAlignments(file = un1, param = param)
```








