---
title: "03 Clustering and classifying your cells"
author: "yincy"
date: "2023-03-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load data 
Single-cell experiments are often performed on tissues containing many cell types. Monocle 3 provides a simple set of functions you can use to group your cells according to their gene expression profiles into clusters. Often cells form clusters that correspond to one cell type or a set of highly related cell types. Monocle 3 uses techniques to do this that are widely accepted in single-cell RNA-seq analysis and similar to the approaches used by `Seurat`, `scanpy`, and other tools.

In this section, you will learn how to cluster cells using Monocle 3. We will demonstrate the main functions used for clustering with the C. elegans data from Cao & Packer et al. This study described how to do single-cell RNA-seq with combinatorial indexing in a protocol called "sci-RNA-seq". Cao & Packer et al. used sci-RNA-seq to produce the first single-cell RNA-seq analysis of a whole animal, so there are many cell types represented in the data. You can learn more about the dataset and see how the authors performed the original analysis at the UW Genome Sciences RNA Atlas of the Worm site.

```{r}
library(monocle3)
library(dplyr)

expression_matrix <- readRDS("../../../data/monocle3/cao_l2_expression.rds")
cell_metadata <- readRDS("../../../data/monocle3/cao_l2_colData.rds")
gene_annotation <- readRDS("../../../data/monocle3/cao_l2_rowData.rds")

cds <- new_cell_data_set(expression_data = expression_matrix, 
                         cell_metadata = cell_metadata, 
                         gene_metadata = gene_annotation)
```


# Pre-process the data
Now that the data's all loaded up, we need to pre-process it. This step is where you tell Monocle 3 how you want to normalize the data, whether to use Principal Components Analysis (the standard for RNA-seq) or Latent Semantic Indexing (common in ATAC-seq), and how to remove any batch effects. We will just use the standard PCA method in this demonstration. When using PCA, you should specify the number of principal components you want Monocle to compute.

```{r}
cds <- preprocess_cds(cds = cds, num_dim = 100)
```

It's a good idea to check that you're using enough PCs to capture most of the variation in gene expression across all the cells in the data set. You can look at the fraction of variation explained by each PC using `plot_pc_variance_explained()`:

```{r}
plot_pc_variance_explained(cds)
```

We can see that using more than 100 PCs would capture only a small amount of additional variation, and each additional PC makes downstream steps in Monocle slower.

# Reduce dimensionality and visualize the cells
Now we're ready to visualize the cells. To do so, you can use either t-SNE, which is very popular in single-cell RNA-seq, or UMAP, which is increasingly common. Monocle 3 uses UMAP by default, as we feel that it is both faster and better suited for clustering and trajectory analysis in RNA-seq. To reduce the dimensionality of the data down into the X, Y plane so we can plot it easily, call `reduce_dimension()`:

```{r}
cds <- reduce_dimension(cds = cds, 
                        max_components = 2, 
                        reduction_method = "UMAP")
```

To plot the data, use Monocle's main plotting function, `plot_cells()`.  

```{r}
plot_cells(cds)
```

Each point in the plot above represents a different cell in the `cell_data_set` object `cds`. As you can see the cells form many groups, some with thousands of cells, some with only a few. Cao & Packer annotated each cell according to type manually by looking at which genes it expresses. We can color the cells in the UMAP plot by the authors' original annotations using the `color_cells_by` argument to `plot_cells()`.

```{r}
plot_cells(cds, color_cells_by = "cao_cell_type")
```

You can see that many of the cell types land very close to one another in the UMAP plot.

Except for a few cases described in a moment, `color_cells_by` can be the name of any column in `colData(cds)`. Note that when `color_cells_by` is a categorical variable, labels are added to the plot, with each label positioned roughly in the middle of all the cells that have that label.

You can also color your cells according to how much of a gene or **set of genes** they express:

```{r}
plot_cells(cds, 
           genes = c("cpna-2", "egl-21", "ram-2", "ions-1"))
```


# Check for and remove batch effects
When performing gene expression analysis, it's important to check for batch effects, which are systematic differences in the transcriptome of cells measured in different experimental batches. These could be technical in nature, such as those introduced during the single-cell RNA-seq protocol, or biological, such as those that might arise from different litters of mice. How to recognize batch effects and account for them so that they don't confound your analysis can be a complex issue, but Monocle provides tools for dealing with them.

You should always check for batch effects when you perform dimensionality reduction. You should add a column to the `colData` that encodes which batch each cell is from. Then you can simply color the cells by batch. Cao & Packer et al included a "plate" annotation in their data, which specifies which sci-RNA-seq plate each cell originated from. Coloring the UMAP by plate reveals:

```{r}
plot_cells(cds, color_cells_by = "plate", label_cell_groups = F)
```

Dramatic batch effects are not evident in this data. If the data contained more substantial variation due to plate, we'd expect to see groups of cells that really only come from one plate. Nevertheless, we can try and remove what batch effect is by running the align_cds() function:

```{r}
cds <- align_cds(cds, num_dim = 100, alignment_group = "plate")
cds <- preprocess_cds(cds, method = "PCA")
cds <- reduce_dimension(cds, reduction_method = "UMAP")
plot_cells(cds, color_cells_by = "plate", label_cell_group = F)
```

When run with the `alignment_group` argument, `align_cds()` tries to remove batch effects using mutual nearest neighbor alignment, a technique introduced by John Marioni's lab. Monocle 3 does so by calling Aaron Lun's excellent package `batchelor`. If you use `align_cds()`, be sure to call `get_citations()` to see how you should cite the software on which Monocle depends.


# Group cells into clusters
Grouping cells into clusters is an important step in identifying the cell types represented in your data. Monocle uses a technique called community detection to group cells. This approach was introduced by Levine et al as part of the phenoGraph algorithm.

```{r}
cds <- cluster_cells(cds, resolution = 1e-5)
plot_cells(cds = cds)
```

Note that now when we call `plot_cells()` with no arguments, it colors the cells by cluster according to default.

The `cluster_cells()` also divides the cells into larger, more well separated groups called partitions, using a statistical test from Alex Wolf et al, introduced as part of their PAGA algorithm. You can visualize these partitions like this:


```{r}
plot_cells(cds, color_cells_by = "partition", group_cells_by = "partition")
```

Once you run `cluster_cells()`, the `plot_cells()` function will label each cluster of cells is labeled separately according to how you want to color the cells. For example, the call below colors the cells according to their cell type annotation, and each cluster is labeled according the most common annotation within it:

```{r}
plot_cells(cds, color_cells_by = "cao_cell_type")
```

You can choose to label whole partitions instead of clusters by passing `group_cells_by="partition"`. You can also plot the top 2 labels per cluster by passing `labels_per_group=2` to `plot_cells()`. Finally, you can disable this labeling policy, making `plot_cells()` behave like it did before we called `cluster_cells()`, like this:


```{r}
plot_cells(cds, color_cells_by = "cao_cell_type", label_groups_by_cluster = F)
```


# Find marker genes expressed by each cluster
Once cells have been clustered, we can ask what genes makes them different from one another. To do that, start by calling the `top_markers()` function:

```{r}
library(dplyr)
marker_test_res <- top_markers(cds = cds, 
                               group_cells_by = "partition", 
                               reference_cells = 1000, 
                               cores = 1)

top_markers <- marker_test_res %>% 
    filter(fraction_expressing >= 0.1) %>% 
    group_by(cell_group) %>% 
    top_n(1, wt = pseudo_R2)

top_markers
```

Now, we can plot the expression and fraction of cells that express each marker in each group with the `plot_genes_by_group` function:

```{r}
plot_genes_by_group(cds, 
                    markers = top_markers %>% pull(gene_id) %>% unique(), 
                    group_cells_by = "partition", 
                    ordering = "maximal_on_diag", 
                    max.size = 3)
```



















